// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/UploadFilePickerTab/UploadFilePickerTab.module.scss.js
require("./UploadFilePickerTab.module.css");
const styles = {
    filePicker: 'filePicker_3ae00659',
    tabsContainer: 'tabsContainer_3ae00659',
    tabContainer: 'tabContainer_3ae00659',
    nav: 'nav_3ae00659',
    tabHeaderContainer: 'tabHeaderContainer_3ae00659',
    breadcrumbNav: 'breadcrumbNav_3ae00659',
    breadcrumbNavItem: 'breadcrumbNavItem_3ae00659',
    tab: 'tab_3ae00659',
    tabOffset: 'tabOffset_3ae00659',
    tabFiles: 'tabFiles_3ae00659',
    scrollablePaneWrapper: 'scrollablePaneWrapper_3ae00659',
    actionButtonsContainer: 'actionButtonsContainer_3ae00659',
    actionButtons: 'actionButtons_3ae00659',
    actionButton: 'actionButton_3ae00659',
    tabHeader: 'tabHeader_3ae00659',
    selectedTab: 'selectedTab_3ae00659',
    header: 'header_3ae00659',
    xlg: 'xlg_3ae00659',
    xxlg: 'xxlg_3ae00659',
    xxxlg: 'xxxlg_3ae00659',
    xxxxlg: 'xxxxlg_3ae00659',
    itemPickerTopBar: 'itemPickerTopBar_3ae00659',
    commandBarNoChevron: 'commandBarNoChevron_3ae00659',
    fileItem: 'fileItem_3ae00659',
    folderItem: 'folderItem_3ae00659',
    fileTypeIcon: 'fileTypeIcon_3ae00659',
    fileTypeIconIcon: 'fileTypeIconIcon_3ae00659',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_3ae00659',
    iconColumnHeader: 'iconColumnHeader_3ae00659',
    fileRow: 'fileRow_3ae00659',
    folderRow: 'folderRow_3ae00659',
    emptyFolder: 'emptyFolder_3ae00659',
    emptyFolderImage: 'emptyFolderImage_3ae00659',
    emptyFolderImageTag: 'emptyFolderImageTag_3ae00659',
    emptyFolderTitle: 'emptyFolderTitle_3ae00659',
    emptyFolderSubText: 'emptyFolderSubText_3ae00659',
    emptyFolderMobile: 'emptyFolderMobile_3ae00659',
    emptyFolderPc: 'emptyFolderPc_3ae00659',
    emptyFolderIcon: 'emptyFolderIcon_3ae00659',
    localTabSinglePreview: 'localTabSinglePreview_3ae00659',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_3ae00659',
    localTabFilename: 'localTabFilename_3ae00659',
    localTabLabel: 'localTabLabel_3ae00659',
    localTabInput: 'localTabInput_3ae00659'
};
export default styles;
