// @microsoft/spfx-heft-plugins/lib-commonjs/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    CONFIG_JSON_SYMBOL: function() {
        return _WebpackConfigurationGenerator.CONFIG_JSON_SYMBOL;
    },
    DEFAULT_PORT: function() {
        return _constants.DEFAULT_PORT;
    },
    DEFAULT_TEMP_FOLDER: function() {
        return _constants.DEFAULT_TEMP_FOLDER;
    },
    DevDeployHelper: function() {
        return _DevDeployHelper.DevDeployHelper;
    },
    DevDeployHooks: function() {
        return _DevDeployPlugin.DevDeployHooks;
    },
    FILE_LOADER_EXTENSIONS: function() {
        return _WebpackConfigurationGenerator.FILE_LOADER_EXTENSIONS;
    },
    IS_SPFX_WEBPACK_CONFIGURATION_SYMBOL: function() {
        return _WebpackConfigurationGenerator.IS_SPFX_WEBPACK_CONFIGURATION_SYMBOL;
    },
    PluginNames: function() {
        return _PluginNames;
    },
    SpDevDeploy: function() {
        return _SpDevDeploy.SpDevDeploy;
    },
    WebpackConfigurationGenerator: function() {
        return _WebpackConfigurationGenerator.WebpackConfigurationGenerator;
    },
    _AzureUploader: function() {
        return _AzureUploader.AzureUploader;
    },
    _COMMON_RULE_OPTIONS: function() {
        return _WebpackConfigurationGenerator.COMMON_RULE_OPTIONS;
    },
    _CumulativeManifestProcessor: function() {
        return _CumulativeManifestProcessor.CumulativeManifestProcessor;
    },
    _CustomizeWebpackConfigurationPlugin: function() {
        return _CustomizeWebpackConfigurationPlugin.default;
    },
    _DependencyDiscoveryMode: function() {
        return _CumulativeManifestProcessor.DependencyDiscoveryMode;
    },
    _DeployAzureStoragePlugin: function() {
        return _DeployAzureStoragePlugin.default;
    },
    _FILE_LOADER_RULE_TEST: function() {
        return _WebpackConfigurationGenerator.FILE_LOADER_RULE_TEST;
    },
    _LegacyExternals: function() {
        return _LegacyExternals.LegacyExternals;
    },
    _LocalizedManifestModuleProcessor: function() {
        return _LocalizedManifestModuleProcessor.LocalizedManifestModuleProcessor;
    },
    _ManifestPlugin: function() {
        return _ManifestPlugin.ManifestPlugin;
    },
    _ServeInfoPlugin: function() {
        return _ServeInfoPlugin.ServeInfoPlugin;
    },
    _SolutionPackager: function() {
        return _SolutionPackager.SolutionPackager;
    },
    _VariableNameGenerator: function() {
        return _VariableNameGenerator.VariableNameGenerator;
    },
    _WebpackStatsPlugin: function() {
        return _WebpackStatsPlugin.WebpackStatsPlugin;
    },
    _WebpackTypes: function() {
        return _webpackTypes;
    },
    _computeIntegrity: function() {
        return _ManifestPlugin.computeIntegrity;
    },
    _getFullHash: function() {
        return _getFullHash.getFullHash;
    },
    _getVersionFromPackageJson: function() {
        return _ExternalsProcessor.getVersionFromPackageJson;
    },
    _includeNonStandardExternal: function() {
        return _ExternalsProcessor.includeNonStandardExternal;
    },
    _loadServeSchemaJson: function() {
        return _serveJsonReader.loadServeSchemaJson;
    },
    _migrateToLatestVersion: function() {
        return _configJsonMigrator.migrateToLatestVersion;
    },
    _normalizeLocaleName: function() {
        return _normalizeLocaleName.normalizeLocaleName;
    },
    _readConfigFileAsync: function() {
        return _configJsonReader.readConfigFileAsync;
    },
    _readServeConfigFile: function() {
        return _serveJsonReader.readServeConfigFile;
    }
});
const _SpDevDeploy = require("./plugins/devDeployPlugin/SpDevDeploy");
const _DevDeployPlugin = require("./plugins/devDeployPlugin/DevDeployPlugin");
const _DevDeployHelper = require("./plugins/devDeployPlugin/DevDeployHelper");
const _PluginNames = /*#__PURE__*/ _interop_require_wildcard(require("./plugins/PluginNames"));
const _WebpackConfigurationGenerator = require("./plugins/webpackConfigurationPlugin/WebpackConfigurationGenerator");
const _WebpackStatsPlugin = require("./plugins/webpackConfigurationPlugin/webpackPlugins/WebpackStatsPlugin");
const _AzureUploader = require("./utilities/AzureUploader");
const _configJsonReader = require("./spfxConfig/configJsonReader");
const _configJsonMigrator = require("./spfxConfig/configJsonMigrator");
const _constants = require("./utilities/constants");
const _normalizeLocaleName = require("./utilities/normalizeLocaleName");
const _serveJsonReader = require("./spfxConfig/serveJsonReader");
const _CumulativeManifestProcessor = require("./spfxManifests/cumulativeManifestProcessor/CumulativeManifestProcessor");
const _getFullHash = require("./spfxManifests/utilities/getFullHash");
const _ExternalsProcessor = require("./spfxManifests/utilities/ExternalsProcessor");
const _ManifestPlugin = require("./spfxManifests/webpack/ManifestPlugin");
const _LocalizedManifestModuleProcessor = require("./spfxManifests/webpack/LocalizedManifestModuleProcessor");
const _CustomizeWebpackConfigurationPlugin = /*#__PURE__*/ _interop_require_default(require("./plugins/CustomizeWebpackConfigurationPlugin"));
const _ServeInfoPlugin = require("./plugins/webpackConfigurationPlugin/webpackPlugins/ServeInfoPlugin");
const _SolutionPackager = require("./plugins/packageSolutionPlugin/SolutionPackager");
const _LegacyExternals = require("./plugins/thirdPartyExternalsPlugin/LegacyExternals");
const _DeployAzureStoragePlugin = /*#__PURE__*/ _interop_require_default(require("./plugins/deployAzureStorage/DeployAzureStoragePlugin"));
const _webpackTypes = /*#__PURE__*/ _interop_require_wildcard(require("./plugins/webpackConfigurationPlugin/webpackPlugins/webpackTypes"));
const _VariableNameGenerator = require("./utilities/VariableNameGenerator");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}

//#sourceMappingUrl=./index.js.map
