// @microsoft/rush-lib/lib-commonjs/utilities/PathConstants.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.scriptsFolderPath = exports.installRunRushPnpmScriptFilename = exports.installRunRushxScriptFilename = exports.installRunRushScriptFilename = exports.installRunScriptFilename = exports.subspacePnpmfileShimFilename = exports.pnpmfileShimFilename = exports.scriptsFolderName = exports.assetsFolderPath = exports.rushLibFolderRootPath = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
/**
 * The currently-executing rush-lib package's root folder path.
 */
exports.rushLibFolderRootPath = node_core_library_1.PackageJsonLookup.instance.tryGetPackageFolderFor(__dirname);
/**
 * The path to the assets folder in rush-lib.
 */
exports.assetsFolderPath = `${exports.rushLibFolderRootPath}/assets`;
/**
 * The folder name ("scripts") where the scripts in rush-lib are built.
 */
exports.scriptsFolderName = 'scripts';
exports.pnpmfileShimFilename = 'PnpmfileShim.js';
exports.subspacePnpmfileShimFilename = 'SubspaceGlobalPnpmfileShim.js';
exports.installRunScriptFilename = 'install-run.js';
exports.installRunRushScriptFilename = 'install-run-rush.js';
exports.installRunRushxScriptFilename = 'install-run-rushx.js';
exports.installRunRushPnpmScriptFilename = 'install-run-rush-pnpm.js';
/**
 * The path to the scripts folder in rush-lib/dist.
 */
exports.scriptsFolderPath = `${exports.rushLibFolderRootPath}/dist/${exports.scriptsFolderName}`;
