// @rushstack/webpack5-localization-plugin/lib/utilities/Constants.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.JSONP_PLACEHOLDER = exports.LOCALE_NAME_PLACEHOLDER = exports.JSONP_PLACEHOLDER_LABEL = exports.LOCALE_NAME_PLACEHOLDER_LABEL = exports.STRING_PLACEHOLDER_LABEL = exports.RESOURCE_FILE_NAME_REGEXP = exports.STRING_PLACEHOLDER_PREFIX = exports.NO_LOCALE_SOURCE_MAP_FILENAME_TOKEN_REGEX = exports.NO_LOCALE_SOURCE_MAP_FILENAME_TOKEN = exports.LOCALE_FILENAME_TOKEN_REGEX = exports.LOCALE_FILENAME_TOKEN = void 0;
exports.LOCALE_FILENAME_TOKEN = '[locale]';
exports.LOCALE_FILENAME_TOKEN_REGEX = /\[locale\]/gi;
exports.NO_LOCALE_SOURCE_MAP_FILENAME_TOKEN = '[no-locale-file]';
exports.NO_LOCALE_SOURCE_MAP_FILENAME_TOKEN_REGEX = /\[no-locale-file\]/gi;
exports.STRING_PLACEHOLDER_PREFIX = '_LOCALIZED_STRING_f12dy0i7_n4bo_dqwj_39gf_sasqehjmihz9';
exports.RESOURCE_FILE_NAME_REGEXP = /\.(resx|resx\.json|loc\.json|resjson)$/i;
exports.STRING_PLACEHOLDER_LABEL = 'A';
exports.LOCALE_NAME_PLACEHOLDER_LABEL = 'B';
exports.JSONP_PLACEHOLDER_LABEL = 'C';
// _LOCALIZED_STRING_f12dy0i7_n4bo_dqwj_39gf_sasqehjmihz9_B_
exports.LOCALE_NAME_PLACEHOLDER = `${exports.STRING_PLACEHOLDER_PREFIX}_${exports.LOCALE_NAME_PLACEHOLDER_LABEL}_`;
// _LOCALIZED_STRING_f12dy0i7_n4bo_dqwj_39gf_sasqehjmihz9_C_
exports.JSONP_PLACEHOLDER = `${exports.STRING_PLACEHOLDER_PREFIX}_${exports.JSONP_PLACEHOLDER_LABEL}_`;
