// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/WebSearchTab/WebSearchTab.module.scss.js
require("./WebSearchTab.module.css");
const styles = {
    filePicker: 'filePicker_ea51e5aa',
    tabsContainer: 'tabsContainer_ea51e5aa',
    tabContainer: 'tabContainer_ea51e5aa',
    nav: 'nav_ea51e5aa',
    tabHeaderContainer: 'tabHeaderContainer_ea51e5aa',
    breadcrumbNav: 'breadcrumbNav_ea51e5aa',
    breadcrumbNavItem: 'breadcrumbNavItem_ea51e5aa',
    tab: 'tab_ea51e5aa',
    tabOffset: 'tabOffset_ea51e5aa',
    tabFiles: 'tabFiles_ea51e5aa',
    scrollablePaneWrapper: 'scrollablePaneWrapper_ea51e5aa',
    actionButtonsContainer: 'actionButtonsContainer_ea51e5aa',
    actionButtons: 'actionButtons_ea51e5aa',
    actionButton: 'actionButton_ea51e5aa',
    tabHeader: 'tabHeader_ea51e5aa',
    selectedTab: 'selectedTab_ea51e5aa',
    header: 'header_ea51e5aa',
    xlg: 'xlg_ea51e5aa',
    xxlg: 'xxlg_ea51e5aa',
    xxxlg: 'xxxlg_ea51e5aa',
    xxxxlg: 'xxxxlg_ea51e5aa',
    itemPickerTopBar: 'itemPickerTopBar_ea51e5aa',
    commandBarNoChevron: 'commandBarNoChevron_ea51e5aa',
    fileItem: 'fileItem_ea51e5aa',
    folderItem: 'folderItem_ea51e5aa',
    fileTypeIcon: 'fileTypeIcon_ea51e5aa',
    fileTypeIconIcon: 'fileTypeIconIcon_ea51e5aa',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_ea51e5aa',
    iconColumnHeader: 'iconColumnHeader_ea51e5aa',
    fileRow: 'fileRow_ea51e5aa',
    folderRow: 'folderRow_ea51e5aa',
    emptyFolder: 'emptyFolder_ea51e5aa',
    emptyFolderImage: 'emptyFolderImage_ea51e5aa',
    emptyFolderImageTag: 'emptyFolderImageTag_ea51e5aa',
    emptyFolderTitle: 'emptyFolderTitle_ea51e5aa',
    emptyFolderSubText: 'emptyFolderSubText_ea51e5aa',
    emptyFolderMobile: 'emptyFolderMobile_ea51e5aa',
    emptyFolderPc: 'emptyFolderPc_ea51e5aa',
    emptyFolderIcon: 'emptyFolderIcon_ea51e5aa',
    searchBoxContainer: 'searchBoxContainer_ea51e5aa',
    searchBoxMedium: 'searchBoxMedium_ea51e5aa',
    searchBox: 'searchBox_ea51e5aa',
    dropdownContainer: 'dropdownContainer_ea51e5aa',
    filterDropdown: 'filterDropdown_ea51e5aa',
    filePickerFolderCardGrid: 'filePickerFolderCardGrid_ea51e5aa',
    filePickerFolderCardTile: 'filePickerFolderCardTile_ea51e5aa',
    filePickerFolderCardPadder: 'filePickerFolderCardPadder_ea51e5aa',
    filePickerFolderCardSizer: 'filePickerFolderCardSizer_ea51e5aa',
    filePickerFolderCardImage: 'filePickerFolderCardImage_ea51e5aa',
    filePickerFolderCardLabel: 'filePickerFolderCardLabel_ea51e5aa',
    bingGridListCell: 'bingGridListCell_ea51e5aa',
    bingTile: 'bingTile_ea51e5aa',
    bingTileContent: 'bingTileContent_ea51e5aa',
    bingTileFrame: 'bingTileFrame_ea51e5aa',
    bingTileCheckCircle: 'bingTileCheckCircle_ea51e5aa',
    bingTileNamePlate: 'bingTileNamePlate_ea51e5aa',
    isSelected: 'isSelected_ea51e5aa',
    copyrightLabel: 'copyrightLabel_ea51e5aa',
    noResultLabel: 'noResultLabel_ea51e5aa',
    bingTileThumbnail: 'bingTileThumbnail_ea51e5aa',
    bingGrildList: 'bingGrildList_ea51e5aa'
};
export default styles;
