// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/FolderTile/FolderTile.module.scss.js
require("./FolderTile.module.css");
const styles = {
    filePicker: 'filePicker_7cebb224',
    tabsContainer: 'tabsContainer_7cebb224',
    tabContainer: 'tabContainer_7cebb224',
    nav: 'nav_7cebb224',
    tabHeaderContainer: 'tabHeaderContainer_7cebb224',
    breadcrumbNav: 'breadcrumbNav_7cebb224',
    breadcrumbNavItem: 'breadcrumbNavItem_7cebb224',
    tab: 'tab_7cebb224',
    tabOffset: 'tabOffset_7cebb224',
    tabFiles: 'tabFiles_7cebb224',
    scrollablePaneWrapper: 'scrollablePaneWrapper_7cebb224',
    actionButtonsContainer: 'actionButtonsContainer_7cebb224',
    actionButtons: 'actionButtons_7cebb224',
    actionButton: 'actionButton_7cebb224',
    tabHeader: 'tabHeader_7cebb224',
    selectedTab: 'selectedTab_7cebb224',
    header: 'header_7cebb224',
    xlg: 'xlg_7cebb224',
    xxlg: 'xxlg_7cebb224',
    xxxlg: 'xxxlg_7cebb224',
    xxxxlg: 'xxxxlg_7cebb224',
    itemPickerTopBar: 'itemPickerTopBar_7cebb224',
    commandBarNoChevron: 'commandBarNoChevron_7cebb224',
    fileItem: 'fileItem_7cebb224',
    folderItem: 'folderItem_7cebb224',
    fileTypeIcon: 'fileTypeIcon_7cebb224',
    fileTypeIconIcon: 'fileTypeIconIcon_7cebb224',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_7cebb224',
    iconColumnHeader: 'iconColumnHeader_7cebb224',
    fileRow: 'fileRow_7cebb224',
    folderRow: 'folderRow_7cebb224',
    emptyFolder: 'emptyFolder_7cebb224',
    emptyFolderImage: 'emptyFolderImage_7cebb224',
    emptyFolderImageTag: 'emptyFolderImageTag_7cebb224',
    emptyFolderTitle: 'emptyFolderTitle_7cebb224',
    emptyFolderSubText: 'emptyFolderSubText_7cebb224',
    emptyFolderMobile: 'emptyFolderMobile_7cebb224',
    emptyFolderPc: 'emptyFolderPc_7cebb224',
    emptyFolderIcon: 'emptyFolderIcon_7cebb224',
    aboveNameplate: 'aboveNameplate_7cebb224',
    tileContent: 'tileContent_7cebb224',
    foreground: 'foreground_7cebb224',
    grid: 'grid_7cebb224',
    cellContent: 'cellContent_7cebb224',
    content: 'content_7cebb224',
    tile: 'tile_7cebb224',
    invokable: 'invokable_7cebb224',
    namePlate: 'namePlate_7cebb224',
    nameplate: 'nameplate_7cebb224',
    name: 'name_7cebb224',
    isLarge: 'isLarge_7cebb224',
    activity: 'activity_7cebb224',
    signalField: 'signalField_7cebb224',
    compact: 'compact_7cebb224',
    signalFieldValue: 'signalFieldValue_7cebb224',
    folderList: 'folderList_7cebb224',
    link: 'link_7cebb224',
    cell: 'cell_7cebb224',
    listCell: 'listCell_7cebb224',
    label: 'label_7cebb224',
    check: 'check_7cebb224',
    selectable: 'selectable_7cebb224',
    showCheck: 'showCheck_7cebb224',
    selected: 'selected_7cebb224',
    dialogMainOverride: 'dialogMainOverride_7cebb224',
    itemTileBlankCover: 'itemTileBlankCover_7cebb224',
    metadata: 'metadata_7cebb224',
    isSmall: 'isSmall_7cebb224',
    listPage: 'listPage_7cebb224',
    odItemTile2FolderCover: 'odItemTile2FolderCover_7cebb224',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_7cebb224',
    folderCover: 'folderCover_7cebb224',
    isSmall_88d29bd4: 'isSmall_88d29bd4_7cebb224',
    folderCoverBack: 'folderCoverBack_7cebb224',
    folderCoverFront: 'folderCoverFront_7cebb224',
    folderCoverFrame: 'folderCoverFrame_7cebb224',
    folderCoverContent: 'folderCoverContent_7cebb224'
};
export default styles;
