// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/StockImagesTab/StockImages.module.scss.js
require("./StockImages.module.css");
const styles = {
    filePicker: 'filePicker_819abf6d',
    tabsContainer: 'tabsContainer_819abf6d',
    tabContainer: 'tabContainer_819abf6d',
    nav: 'nav_819abf6d',
    tabHeaderContainer: 'tabHeaderContainer_819abf6d',
    breadcrumbNav: 'breadcrumbNav_819abf6d',
    breadcrumbNavItem: 'breadcrumbNavItem_819abf6d',
    tab: 'tab_819abf6d',
    tabOffset: 'tabOffset_819abf6d',
    tabFiles: 'tabFiles_819abf6d',
    scrollablePaneWrapper: 'scrollablePaneWrapper_819abf6d',
    actionButtonsContainer: 'actionButtonsContainer_819abf6d',
    actionButtons: 'actionButtons_819abf6d',
    actionButton: 'actionButton_819abf6d',
    tabHeader: 'tabHeader_819abf6d',
    selectedTab: 'selectedTab_819abf6d',
    header: 'header_819abf6d',
    xlg: 'xlg_819abf6d',
    xxlg: 'xxlg_819abf6d',
    xxxlg: 'xxxlg_819abf6d',
    xxxxlg: 'xxxxlg_819abf6d',
    itemPickerTopBar: 'itemPickerTopBar_819abf6d',
    commandBarNoChevron: 'commandBarNoChevron_819abf6d',
    fileItem: 'fileItem_819abf6d',
    folderItem: 'folderItem_819abf6d',
    fileTypeIcon: 'fileTypeIcon_819abf6d',
    fileTypeIconIcon: 'fileTypeIconIcon_819abf6d',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_819abf6d',
    iconColumnHeader: 'iconColumnHeader_819abf6d',
    fileRow: 'fileRow_819abf6d',
    folderRow: 'folderRow_819abf6d',
    emptyFolder: 'emptyFolder_819abf6d',
    emptyFolderImage: 'emptyFolderImage_819abf6d',
    emptyFolderImageTag: 'emptyFolderImageTag_819abf6d',
    emptyFolderTitle: 'emptyFolderTitle_819abf6d',
    emptyFolderSubText: 'emptyFolderSubText_819abf6d',
    emptyFolderMobile: 'emptyFolderMobile_819abf6d',
    emptyFolderPc: 'emptyFolderPc_819abf6d',
    emptyFolderIcon: 'emptyFolderIcon_819abf6d',
    stockImagesPickerContainer: 'stockImagesPickerContainer_819abf6d',
    stockImagesPicker: 'stockImagesPicker_819abf6d'
};
export default styles;
