// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/LinkFilePickerTab/LinkFilePickerTab.module.scss.js
require("./LinkFilePickerTab.module.css");
const styles = {
    filePicker: 'filePicker_e85f0b4f',
    tabsContainer: 'tabsContainer_e85f0b4f',
    tabContainer: 'tabContainer_e85f0b4f',
    nav: 'nav_e85f0b4f',
    tabHeaderContainer: 'tabHeaderContainer_e85f0b4f',
    breadcrumbNav: 'breadcrumbNav_e85f0b4f',
    breadcrumbNavItem: 'breadcrumbNavItem_e85f0b4f',
    tab: 'tab_e85f0b4f',
    tabOffset: 'tabOffset_e85f0b4f',
    tabFiles: 'tabFiles_e85f0b4f',
    scrollablePaneWrapper: 'scrollablePaneWrapper_e85f0b4f',
    actionButtonsContainer: 'actionButtonsContainer_e85f0b4f',
    actionButtons: 'actionButtons_e85f0b4f',
    actionButton: 'actionButton_e85f0b4f',
    tabHeader: 'tabHeader_e85f0b4f',
    selectedTab: 'selectedTab_e85f0b4f',
    header: 'header_e85f0b4f',
    xlg: 'xlg_e85f0b4f',
    xxlg: 'xxlg_e85f0b4f',
    xxxlg: 'xxxlg_e85f0b4f',
    xxxxlg: 'xxxxlg_e85f0b4f',
    itemPickerTopBar: 'itemPickerTopBar_e85f0b4f',
    commandBarNoChevron: 'commandBarNoChevron_e85f0b4f',
    fileItem: 'fileItem_e85f0b4f',
    folderItem: 'folderItem_e85f0b4f',
    fileTypeIcon: 'fileTypeIcon_e85f0b4f',
    fileTypeIconIcon: 'fileTypeIconIcon_e85f0b4f',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_e85f0b4f',
    iconColumnHeader: 'iconColumnHeader_e85f0b4f',
    fileRow: 'fileRow_e85f0b4f',
    folderRow: 'folderRow_e85f0b4f',
    emptyFolder: 'emptyFolder_e85f0b4f',
    emptyFolderImage: 'emptyFolderImage_e85f0b4f',
    emptyFolderImageTag: 'emptyFolderImageTag_e85f0b4f',
    emptyFolderTitle: 'emptyFolderTitle_e85f0b4f',
    emptyFolderSubText: 'emptyFolderSubText_e85f0b4f',
    emptyFolderMobile: 'emptyFolderMobile_e85f0b4f',
    emptyFolderPc: 'emptyFolderPc_e85f0b4f',
    emptyFolderIcon: 'emptyFolderIcon_e85f0b4f',
    linkTextField: 'linkTextField_e85f0b4f',
    localTabSinglePreview: 'localTabSinglePreview_e85f0b4f',
    localTabSinglePreviewImage: 'localTabSinglePreviewImage_e85f0b4f'
};
export default styles;
