// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/OneDriveFilesTab/OneDriveFilesTab.module.scss.js
require("./OneDriveFilesTab.module.css");
const styles = {
    filePicker: 'filePicker_2d411467',
    tabsContainer: 'tabsContainer_2d411467',
    tabContainer: 'tabContainer_2d411467',
    nav: 'nav_2d411467',
    tabHeaderContainer: 'tabHeaderContainer_2d411467',
    breadcrumbNav: 'breadcrumbNav_2d411467',
    breadcrumbNavItem: 'breadcrumbNavItem_2d411467',
    tab: 'tab_2d411467',
    tabOffset: 'tabOffset_2d411467',
    tabFiles: 'tabFiles_2d411467',
    scrollablePaneWrapper: 'scrollablePaneWrapper_2d411467',
    actionButtonsContainer: 'actionButtonsContainer_2d411467',
    actionButtons: 'actionButtons_2d411467',
    actionButton: 'actionButton_2d411467',
    tabHeader: 'tabHeader_2d411467',
    selectedTab: 'selectedTab_2d411467',
    header: 'header_2d411467',
    xlg: 'xlg_2d411467',
    xxlg: 'xxlg_2d411467',
    xxxlg: 'xxxlg_2d411467',
    xxxxlg: 'xxxxlg_2d411467',
    itemPickerTopBar: 'itemPickerTopBar_2d411467',
    commandBarNoChevron: 'commandBarNoChevron_2d411467',
    fileItem: 'fileItem_2d411467',
    folderItem: 'folderItem_2d411467',
    fileTypeIcon: 'fileTypeIcon_2d411467',
    fileTypeIconIcon: 'fileTypeIconIcon_2d411467',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_2d411467',
    iconColumnHeader: 'iconColumnHeader_2d411467',
    fileRow: 'fileRow_2d411467',
    folderRow: 'folderRow_2d411467',
    emptyFolder: 'emptyFolder_2d411467',
    emptyFolderImage: 'emptyFolderImage_2d411467',
    emptyFolderImageTag: 'emptyFolderImageTag_2d411467',
    emptyFolderTitle: 'emptyFolderTitle_2d411467',
    emptyFolderSubText: 'emptyFolderSubText_2d411467',
    emptyFolderMobile: 'emptyFolderMobile_2d411467',
    emptyFolderPc: 'emptyFolderPc_2d411467',
    emptyFolderIcon: 'emptyFolderIcon_2d411467'
};
export default styles;
