// @microsoft/sp-css-loader/lib-commonjs/index.js
/**
 * Inspired by css-loader
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, /**
 * @public
 */ "default", {
    enumerable: true,
    get: function() {
        return loaderFn;
    }
});
const _path = /*#__PURE__*/ _interop_require_wildcard(require("path"));
const _loaderutils = /*#__PURE__*/ _interop_require_wildcard(require("loader-utils"));
const _plugins = /*#__PURE__*/ _interop_require_wildcard(require("css-loader/dist/plugins"));
const _Warning = /*#__PURE__*/ _interop_require_default(require("css-loader/dist/Warning"));
const _CssSyntaxError = /*#__PURE__*/ _interop_require_default(require("css-loader/dist/CssSyntaxError"));
const _postcss = /*#__PURE__*/ _interop_require_default(require("postcss"));
const _cssnano = /*#__PURE__*/ _interop_require_default(require("cssnano"));
const _postcssmodulesvalues = /*#__PURE__*/ _interop_require_default(require("postcss-modules-values"));
const _postcssmoduleslocalbydefault = /*#__PURE__*/ _interop_require_default(require("postcss-modules-local-by-default"));
const _postcssmodulesextractimports = /*#__PURE__*/ _interop_require_default(require("postcss-modules-extract-imports"));
const _postcssmodulesscope = /*#__PURE__*/ _interop_require_default(require("postcss-modules-scope"));
const _autoprefixer = /*#__PURE__*/ _interop_require_default(require("autoprefixer"));
const _nodecorelibrary = require("@rushstack/node-core-library");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function loaderFn(source) {
    const callback = this.async();
    if (!callback) {
        throw new Error('Plugin is async, but a callback was not provided');
    }
    processCssAsync.call(this, source, this.resourcePath).then((result)=>callback(undefined, result)).catch((error)=>callback(error));
}
const GET_URL_IMPORT_NAME = '___CSS_LOADER_GET_URL_IMPORT___';
const LOAD_THEMED_STYLES_IMPORT_NAME = '__LOAD_THEMED_STYLES__';
async function processCssAsync(content, resourcePath) {
    const options = this.getOptions();
    const processor = getPostcssProcessor(options.generateCssClassName, options.production);
    let result;
    try {
        result = await processor.process(content, {
            from: resourcePath,
            to: resourcePath
        });
    } catch (error) {
        if (error.name === 'CssSyntaxError') {
            throw new _CssSyntaxError.default(error);
        } else {
            throw error;
        }
    }
    for (const warning of result.warnings()){
        this.emitWarning(new _Warning.default(warning));
    }
    const loadThemedStylesModulePath = options.loadThemedStylesImportPath || getLoadThemedStylesModulePath(this.context);
    const importLines = [
        `import * as ${LOAD_THEMED_STYLES_IMPORT_NAME} from ${JSON.stringify(loadThemedStylesModulePath)};`
    ];
    const codeLines = [];
    const exportLines = [];
    let getUrlHasBeenImported = false;
    let urlImportCounter = 0;
    let cssText = JSON.stringify(result.content);
    for (const message of result.messages){
        switch(message.type){
            case 'import':
                {
                    const importMessage = message.value;
                    if (importMessage.type !== 'url') {
                        throw new Error(`Only the "url" import type is supported. Encountered "${importMessage.type}"`);
                    }
                    if (!getUrlHasBeenImported) {
                        const getUrlModulePath = getGetUrlModulePath(this.context);
                        importLines.push(`import ${GET_URL_IMPORT_NAME} from ${JSON.stringify(getUrlModulePath)};`);
                        getUrlHasBeenImported = true;
                    }
                    const importName = `__URL_IMPORT_${urlImportCounter++}__`;
                    importLines.push(`import ${importName} from ${JSON.stringify(importMessage.url)}`);
                    codeLines.push(`var ${importMessage.replacementName} = ${GET_URL_IMPORT_NAME}(${importName})`);
                    break;
                }
            case 'export':
                {
                    const exportMessage = message.value;
                    const exportName = exportMessage.name.match(/^[A-z_][A-z0-9_]*$/) ? exportMessage.name : JSON.stringify(exportMessage.name);
                    exportLines.push(`  ${exportName}: ${JSON.stringify(exportMessage.value)}`);
                    break;
                }
            case 'replacer':
                {
                    const replacerMessage = message.value;
                    if (replacerMessage.type !== 'url') {
                        throw new Error(`Only the "url" replacer type is supported. Encountered "${replacerMessage.type}"`);
                    }
                    cssText = cssText.replace(// eslint-disable-next-line @rushstack/security/no-unsafe-regexp
                    new RegExp(replacerMessage.replacementName, 'g'), `" + ${replacerMessage.replacementName} + "`);
                    break;
                }
            case 'removal':
                {
                    if (message.plugin !== 'postcss-discard-empty') {
                        throw new Error('Unknown source of "removal" message type encountered in simple-css-loader.');
                    }
                    break;
                }
            case 'warning':
                {
                    this.emitWarning(message.value);
                    break;
                }
            default:
                {
                    throw new Error(`Unexpected message type: "${message.type}"`);
                }
        }
    }
    if (options.async) {
        codeLines.push(`${LOAD_THEMED_STYLES_IMPORT_NAME}.loadStyles(${cssText}, true);`);
    } else {
        codeLines.push(`${LOAD_THEMED_STYLES_IMPORT_NAME}.loadStyles(${cssText});`);
    }
    const allLines = [
        '// Imports',
        ...importLines,
        '',
        ...codeLines,
        ''
    ];
    if (exportLines.length > 0) {
        allLines.push('// Exports', 'export default {', exportLines.join(',\n'), '};', '');
    }
    return allLines.join('\n');
}
const PRE_MODULE_PLUGINS = [
    (0, _cssnano.default)({
        preset: 'default'
    }),
    (0, _autoprefixer.default)({
        overrideBrowserslist: [
            '> 1%',
            'ie >= 11'
        ]
    })
];
const POST_MODULE_PLUGINS = [
    _plugins.icssParser({}),
    _plugins.importParser({}),
    _plugins.urlParser({
        filter: (item)=>{
            // Don't try to resolve data URLs
            return _loaderutils.isUrlRequest(item) && !item.startsWith('data:');
        }
    })
];
const POSTCSS_PROCESSOR_CACHE = new Map();
function getPostcssProcessor(generateCssClassName, production) {
    let processor = POSTCSS_PROCESSOR_CACHE.get(generateCssClassName);
    if (!processor) {
        let plugins;
        const localPreModulePlugins = [];
        if (generateCssClassName) {
            plugins = [
                ...PRE_MODULE_PLUGINS,
                ...localPreModulePlugins,
                _postcssmodulesvalues.default,
                (0, _postcssmoduleslocalbydefault.default)({
                    mode: 'local'
                }),
                (0, _postcssmodulesextractimports.default)(),
                (0, _postcssmodulesscope.default)({
                    generateScopedName: (existingClassName, cssFilePath, cssContent)=>{
                        return generateCssClassName(existingClassName, cssFilePath, cssContent, production);
                    }
                }),
                ...POST_MODULE_PLUGINS
            ];
        } else {
            plugins = [
                ...PRE_MODULE_PLUGINS,
                ...localPreModulePlugins,
                ...POST_MODULE_PLUGINS
            ];
        }
        processor = (0, _postcss.default)(plugins);
        POSTCSS_PROCESSOR_CACHE.set(generateCssClassName, processor);
    }
    return processor;
}
const CSS_LOADER_PACKAGE_PATH = _nodecorelibrary.Import.resolvePackage({
    packageName: 'css-loader',
    baseFolderPath: __dirname
});
const CSS_LOADER_RELATIVE_PATH_CACHE = new Map();
function getGetUrlModulePath(context) {
    let relativePath = CSS_LOADER_RELATIVE_PATH_CACHE.get(context);
    if (!relativePath) {
        relativePath = _path.relative(context, CSS_LOADER_PACKAGE_PATH);
        relativePath = relativePath.replace(/\\/g, '/');
        CSS_LOADER_RELATIVE_PATH_CACHE.set(context, relativePath);
    }
    return _path.posix.join(relativePath, 'dist', 'runtime', 'getUrl.js');
}
let LOAD_THEMED_STYLES_PATH;
const LOAD_THEMED_STYLES_RELATIVE_PATH_CACHE = new Map();
function getLoadThemedStylesModulePath(context) {
    let relativePath = LOAD_THEMED_STYLES_RELATIVE_PATH_CACHE.get(context);
    if (!relativePath) {
        if (!LOAD_THEMED_STYLES_PATH) {
            LOAD_THEMED_STYLES_PATH = _nodecorelibrary.Import.resolvePackage({
                packageName: '@microsoft/load-themed-styles',
                baseFolderPath: __dirname
            });
        }
        relativePath = _path.relative(context, LOAD_THEMED_STYLES_PATH);
        relativePath = relativePath.replace(/\\/g, '/');
        LOAD_THEMED_STYLES_RELATIVE_PATH_CACHE.set(context, relativePath);
    }
    return _path.posix.join(relativePath, 'lib-es6', 'index.js');
}

//#sourceMappingUrl=./index.js.map
