// @microsoft/sp-property-pane/lib-commonjs/propertyPaneFields/propertyPaneSortableAccordion/PropertyPaneSortableAccordion.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.useStyles = void 0;
var tslib_1 = require("tslib");
var react_components_1 = require("@fluentui/react-components");
exports.useStyles = (0, react_components_1.makeStyles)({
    accordionItem: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        position: 'relative',
        '::before': {
            content: '""',
            display: 'var(--focusIndicatorDisplay)',
            position: 'absolute',
            backgroundColor: react_components_1.tokens.colorBrandBackground,
            width: '2px',
            height: '100%',
            left: '-20px'
        }
    },
    accordionItemHeader: tslib_1.__assign(tslib_1.__assign({ display: 'flex', flexDirection: 'row', alignItems: 'center', columnGap: '8px' }, react_components_1.shorthands.padding('4px', '0px')), { width: '100%' }),
    accordionItemTitle: {
        flexShrink: 1,
        flexGrow: 1,
        overflow: 'hidden',
        textOverflow: 'ellipsis',
        whiteSpace: 'nowrap'
    },
    accordionItemButtons: {
        flexShrink: 0,
        display: 'flex',
        flexDirection: 'row',
        gap: '10px'
    },
    iconButton: tslib_1.__assign({}, react_components_1.shorthands.border('none')),
    dragButton: tslib_1.__assign({ cursor: 'move', flexShrink: 0 }, react_components_1.shorthands.border('none')),
    itemConfigurePane: tslib_1.__assign({}, react_components_1.shorthands.padding('8px', '8px', '32px', '8px'))
});
