// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/SiteFilePickerTab/SiteFilePickerTab.module.scss.js
require("./SiteFilePickerTab.module.css");
const styles = {
    filePicker: 'filePicker_507046eb',
    tabsContainer: 'tabsContainer_507046eb',
    tabContainer: 'tabContainer_507046eb',
    nav: 'nav_507046eb',
    tabHeaderContainer: 'tabHeaderContainer_507046eb',
    breadcrumbNav: 'breadcrumbNav_507046eb',
    breadcrumbNavItem: 'breadcrumbNavItem_507046eb',
    tab: 'tab_507046eb',
    tabOffset: 'tabOffset_507046eb',
    tabFiles: 'tabFiles_507046eb',
    scrollablePaneWrapper: 'scrollablePaneWrapper_507046eb',
    actionButtonsContainer: 'actionButtonsContainer_507046eb',
    actionButtons: 'actionButtons_507046eb',
    actionButton: 'actionButton_507046eb',
    tabHeader: 'tabHeader_507046eb',
    selectedTab: 'selectedTab_507046eb',
    header: 'header_507046eb',
    xlg: 'xlg_507046eb',
    xxlg: 'xxlg_507046eb',
    xxxlg: 'xxxlg_507046eb',
    xxxxlg: 'xxxxlg_507046eb',
    itemPickerTopBar: 'itemPickerTopBar_507046eb',
    commandBarNoChevron: 'commandBarNoChevron_507046eb',
    fileItem: 'fileItem_507046eb',
    folderItem: 'folderItem_507046eb',
    fileTypeIcon: 'fileTypeIcon_507046eb',
    fileTypeIconIcon: 'fileTypeIconIcon_507046eb',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_507046eb',
    iconColumnHeader: 'iconColumnHeader_507046eb',
    fileRow: 'fileRow_507046eb',
    folderRow: 'folderRow_507046eb',
    emptyFolder: 'emptyFolder_507046eb',
    emptyFolderImage: 'emptyFolderImage_507046eb',
    emptyFolderImageTag: 'emptyFolderImageTag_507046eb',
    emptyFolderTitle: 'emptyFolderTitle_507046eb',
    emptyFolderSubText: 'emptyFolderSubText_507046eb',
    emptyFolderMobile: 'emptyFolderMobile_507046eb',
    emptyFolderPc: 'emptyFolderPc_507046eb',
    emptyFolderIcon: 'emptyFolderIcon_507046eb'
};
export default styles;
