// @microsoft/rush-lib/lib-esnext/logic/setup/ArtifactoryConfiguration.js
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
import { JsonFile, JsonSchema, FileSystem } from '@rushstack/node-core-library';
import schemaJson from '../../schemas/artifactory.schema.json';
/**
 * Use this class to load the "common/config/rush/artifactory.json" config file.
 * It configures the "rush setup" command.
 */
export class ArtifactoryConfiguration {
    /**
     * @internal
     */
    constructor(jsonFileName) {
        this._jsonFileName = jsonFileName;
        this.configuration = {
            packageRegistry: {
                enabled: false,
                registryUrl: '',
                artifactoryWebsiteUrl: ''
            }
        };
        if (FileSystem.exists(this._jsonFileName)) {
            this.configuration = JsonFile.loadAndValidate(this._jsonFileName, ArtifactoryConfiguration._jsonSchema);
            if (!this.configuration.packageRegistry.credentialType) {
                this.configuration.packageRegistry.credentialType = 'password';
            }
        }
    }
}
ArtifactoryConfiguration._jsonSchema = JsonSchema.fromLoadedObject(schemaJson);
