// @microsoft/spfx-heft-plugins/lib-esm/plugins/devDeployPlugin/DevDeployHelper.js
import { AzureUploader } from '../../utilities/AzureUploader';
import { UrlUtilities } from '../../utilities/UrlUtilities';
import { ProgressBar } from '../../utilities/ProgressBar';
/**
 * Helper functions for dev-deploy.
 * @public
 */ export class DevDeployHelper {
    static async deployFilesAsync(terminal, deployConfig, deployData, deployPathsFunction) {
        const progressBar = new ProgressBar(deployData.filesToUpload.length);
        let uploadedCount = 0;
        if (deployPathsFunction) {
            await deployPathsFunction(deployData.filesToUpload);
        } else {
            const uploader = new AzureUploader({
                storageAccountName: deployConfig.accountName,
                containerName: deployConfig.containerName,
                sas: deployConfig.sas
            });
            const onFileUploaded = ()=>{
                progressBar.update(++uploadedCount);
            };
            await uploader.uploadFilesToAzureAsync(terminal, deployData.filesToUpload.map((fileToUpload)=>{
                const azurePath = deployConfig.folderNameInContainer ? UrlUtilities.joinUrlSegments(deployConfig.folderNameInContainer, fileToUpload.deployPath) : fileToUpload.deployPath;
                return {
                    azurePath,
                    localPath: fileToUpload.localPath
                };
            }), undefined, deployConfig.maxRetries, undefined, onFileUploaded);
        }
    }
}

//#sourceMappingUrl=./DevDeployHelper.js.map
