// @microsoft/rush-lib/lib-commonjs/logic/operations/NullOperationRunner.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.NullOperationRunner = void 0;
/**
 * Implementation of `IOperationRunner` for operations that require no work, such as empty scripts,
 * skipped operations, or blocked operations.
 */
class NullOperationRunner {
    constructor({ name, result, silent }) {
        // This operation does nothing, so timing is meaningless
        this.reportTiming = false;
        // The operation is a no-op, so it is faster to not cache it
        this.cacheable = false;
        // Nothing will get logged, no point allowing warnings
        this.warningsAreAllowed = false;
        this.isNoOp = true;
        this.name = name;
        this.result = result;
        this.silent = silent;
    }
    async executeAsync(context) {
        return this.result;
    }
    getConfigHash() {
        return '';
    }
}
exports.NullOperationRunner = NullOperationRunner;
