// @microsoft/rush-lib/lib-commonjs/logic/operations/OperationStateFile.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.OperationStateFile = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
/**
 * A helper class for managing the state file of a operation.
 *
 * @internal
 */
class OperationStateFile {
    constructor(options) {
        const { projectFolder, metadataFolder } = options;
        this.relativeFilepath = `${metadataFolder}/${OperationStateFile.filename}`;
        this.filepath = `${projectFolder}/${this.relativeFilepath}`;
    }
    get state() {
        return this._state;
    }
    async writeAsync(json) {
        await node_core_library_1.JsonFile.saveAsync(json, this.filepath, { ensureFolderExists: true, ignoreUndefinedValues: true });
        this._state = json;
    }
    async tryRestoreAsync() {
        try {
            this._state = await node_core_library_1.JsonFile.loadAsync(this.filepath);
        }
        catch (error) {
            if (node_core_library_1.FileSystem.isNotExistError(error)) {
                this._state = undefined;
            }
            else {
                // This should not happen
                throw new node_core_library_1.InternalError(error);
            }
        }
        return this._state;
    }
}
exports.OperationStateFile = OperationStateFile;
OperationStateFile.filename = 'state.json';
