// @microsoft/sp-loader/lib-commonjs/utilities/KillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isFindAppManifestInsteadUsingFirstKSActivated = exports.isPassClientSideApplicationManifestToRootScopeKSActivated = exports.isLoadTrackTimesKSActivated = exports.isLogSPFxVersionFor3pComponentsKSActivated = exports.isSetSameSiteByDefaultKSActivated = exports.notReadTenantIdAndUserIdFromQueryParametersOnFirstPartyIsloatedDomainKSActivated = exports.isPreventLateIframeRedirectKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isPreventLateIframeRedirectKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('56151a3e-a100-403c-9e91-cb5d1677ed66'
    /* '06/13/2022', 'instantiate IframeLoadRedirector earlier and support pre-emptively preventing redirects' */
    );
}
exports.isPreventLateIframeRedirectKSActivated = isPreventLateIframeRedirectKSActivated;
function notReadTenantIdAndUserIdFromQueryParametersOnFirstPartyIsloatedDomainKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('03d09d7b-3e05-4478-a47e-9f91f38574bb'
    /* '08/31/2022', 'kill switch to disable using query parameters for tenant id and user id on securebroker domain' */
    );
}
exports.notReadTenantIdAndUserIdFromQueryParametersOnFirstPartyIsloatedDomainKSActivated = notReadTenantIdAndUserIdFromQueryParametersOnFirstPartyIsloatedDomainKSActivated;
function isSetSameSiteByDefaultKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c6d41898-d9e1-4ec8-9ae5-0c8991c9b5d0'
    /* '02/23/2023, 'Add SameSite attribute be default'  */
    );
}
exports.isSetSameSiteByDefaultKSActivated = isSetSameSiteByDefaultKSActivated;
function isLogSPFxVersionFor3pComponentsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b35cdec2-b17e-473e-8107-ca372f006eea'
    /*'08/31/2022', 'Log SPFx version for 3p components' */
    );
}
exports.isLogSPFxVersionFor3pComponentsKSActivated = isLogSPFxVersionFor3pComponentsKSActivated;
function isLoadTrackTimesKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b90e1867-5423-4e90-ae5b-c38e9faf3e89'
    /*'02/20/2024', 'Track js load times' */
    );
}
exports.isLoadTrackTimesKSActivated = isLoadTrackTimesKSActivated;
function isPassClientSideApplicationManifestToRootScopeKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1e225aae-d8b1-41e7-b8ef-4c736fe61057'
    /*'08/21/2024', 'aterentiev - Pass client side application manifest to root scope' */
    );
}
exports.isPassClientSideApplicationManifestToRootScopeKSActivated = isPassClientSideApplicationManifestToRootScopeKSActivated;
function isFindAppManifestInsteadUsingFirstKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f298c6ca-1f18-4a32-bd13-dc5ed406ec66'
    /*'08/21/2024', 'aterentiev - Find app manifest instead of using first' */
    );
}
exports.isFindAppManifestInsteadUsingFirstKSActivated = isFindAppManifestInsteadUsingFirstKSActivated;
