// @microsoft/rush-lib/lib-esnext/utilities/PathConstants.js
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
import { PackageJsonLookup } from '@rushstack/node-core-library';
/**
 * The currently-executing rush-lib package's root folder path.
 */
export const rushLibFolderRootPath = PackageJsonLookup.instance.tryGetPackageFolderFor(__dirname);
/**
 * The path to the assets folder in rush-lib.
 */
export const assetsFolderPath = `${rushLibFolderRootPath}/assets`;
/**
 * The folder name ("scripts") where the scripts in rush-lib are built.
 */
export const scriptsFolderName = 'scripts';
export const pnpmfileShimFilename = 'PnpmfileShim.js';
export const subspacePnpmfileShimFilename = 'SubspaceGlobalPnpmfileShim.js';
export const installRunScriptFilename = 'install-run.js';
export const installRunRushScriptFilename = 'install-run-rush.js';
export const installRunRushxScriptFilename = 'install-run-rushx.js';
export const installRunRushPnpmScriptFilename = 'install-run-rush-pnpm.js';
/**
 * The path to the scripts folder in rush-lib/dist.
 */
export const scriptsFolderPath = `${rushLibFolderRootPath}/dist/${scriptsFolderName}`;
