// @microsoft/sp-http-base/lib-commonjs/oauthTokenProvider/AadFlights.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isUpgradeMsalVersionEnabled = exports.isUseMsalV3For3pRequestsEnabled = exports.isNaaFlightEnabledForVCM = exports.isNaaFlightEnabledForSharePointMobile = exports.isNaaFlightEnabled = exports.isUseLocalStorageInMsalV3Enabled = exports.isSafariAuthPatchEnbled = exports.isOBOForAllTeamsScenariosEnabled = exports.isAadTokenProviderPopupFlowV1Enabled = exports.isPoPTokenFlightEnabled = exports.isMsalBrowserV3FlightEnabled = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isMsalBrowserV3FlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(61223);
}
exports.isMsalBrowserV3FlightEnabled = isMsalBrowserV3FlightEnabled;
function isPoPTokenFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1512);
}
exports.isPoPTokenFlightEnabled = isPoPTokenFlightEnabled;
function isAadTokenProviderPopupFlowV1Enabled() {
    return sp_core_library_1._SPFlight.isEnabled(1791);
}
exports.isAadTokenProviderPopupFlowV1Enabled = isAadTokenProviderPopupFlowV1Enabled;
/**
 * APEXUseOboForAllTeamsScenarios
 */
function isOBOForAllTeamsScenariosEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(60244);
}
exports.isOBOForAllTeamsScenariosEnabled = isOBOForAllTeamsScenariosEnabled;
/**
 * SppplatSafariAuthPatch
 */
function isSafariAuthPatchEnbled() {
    return sp_core_library_1._SPFlight.isEnabled(687);
}
exports.isSafariAuthPatchEnbled = isSafariAuthPatchEnbled;
/**
 * UseLocalStorageInMsalV3
 */
function isUseLocalStorageInMsalV3Enabled() {
    return sp_core_library_1._SPFlight.isEnabled(61291);
}
exports.isUseLocalStorageInMsalV3Enabled = isUseLocalStorageInMsalV3Enabled;
/**
 * APEXTeamsNaaEnabled
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXTeamsNaaEnabled
 */
function isNaaFlightEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(61858);
}
exports.isNaaFlightEnabled = isNaaFlightEnabled;
/**
 * APEXMobileNaaEnabled - NAA mobile flight for SP Mobile
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXMobileNaaEnabled
 */
function isNaaFlightEnabledForSharePointMobile() {
    return sp_core_library_1._SPFlight.isEnabled(62060);
}
exports.isNaaFlightEnabledForSharePointMobile = isNaaFlightEnabledForSharePointMobile;
/**
 * APEXMobileNaaEnabledVCM - NAA mobile flight for VCM/MeeBridge
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXMobileNaaEnabledVCM
 */
function isNaaFlightEnabledForVCM() {
    return sp_core_library_1._SPFlight.isEnabled(62157);
}
exports.isNaaFlightEnabledForVCM = isNaaFlightEnabledForVCM;
/**
 * APEXUseMsalV3For3PRequests - Use MSALV3 for dev geo requests as the dev geo is configured for V3
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXUseMsalV3For3PRequests
 */
function isUseMsalV3For3pRequestsEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1523);
}
exports.isUseMsalV3For3pRequestsEnabled = isUseMsalV3For3pRequestsEnabled;
/**
 * APEXUpgradeMsalVersion - Update MSAL.js to latest version
 * https://spoflights.azurewebsites.net/Flight/Edit/APEXUpgradeMsalVersion
 */
function isUpgradeMsalVersionEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(62380);
}
exports.isUpgradeMsalVersionEnabled = isUpgradeMsalVersionEnabled;
