// @microsoft/sp-http-base/lib-commonjs/oauthTokenProvider/AadKillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isSkipCacheForSilentAuthKSActivated = exports.isUseUpdatedBrowserUtilitiesVCMCheckKSActivated = exports.isRemoveFloatingPromisesKSActivated = exports.isRetryMonitorWindowTimeoutKSActivated = exports.isUseUpdatedBrowserUtilitiesSPMobileCheckKSActivated = exports.isHandleClaimsChallengeOn1p3pExchangeFailureKSActivated = exports.isLogUserCancelledErrorAsExpectedFailureKSActivated = exports.isUseClientIndexKSActivated = exports.isExpectedResultOnErrorReturnedKSActivated = exports.isAlwaysUseMsalV3KSActivated = exports.isEnsureStateOnNAAKSActivated = exports.isLogCorrelationIdKSActivated = exports.isHandleNAAInteractionRequiredWithPopupKSActivated = exports.isModifyPopupLogicToCallRequestPopupKSActivated = exports.isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated = exports.isEnableEncryptedAuthorizeResponseKSActivated = exports.isEnablePopupFlowForMsalV3KSActivated = exports.isDisableMsalV1KSActivated = exports.isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated = exports.isFailAllEmailAuthGuestTokenRequestsKSActivated = exports.isFailAllAnonymousTokenRequestsKSActivated = exports.isLogAllFullPageRedirectsKSActivated = exports.isConditionalAccessIgnoredOnEnsureStateKSActivated = exports.isUpdateReplyUriForV1FallbackKSActivated = exports.isEnsureRefreshTokenKSActivated = exports.isTenantAdminAppFirtPartyKSActivated = exports.isAlwaysInitOboConfigKSActive = exports.isMsalImplicitV1TokenByPopupKSActivated = exports.isLogAfterFullPageRedirectKSActivated = exports.useOBOInWebForSecureBrokerKSActivated = exports.isTeamsBrowserHostedKSActive = exports.isMsalUseSeparateStorageKeysKsActivated = exports.isMsalBrowserRetryAndSsoSilentImprovementsKsActivated = exports.isMsalTwoMinuteMaxRedirectKsActivated = exports.isManualRedirectCancellationActivated = exports.isSkipKnownErroneousResourcesActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isSkipKnownErroneousResourcesActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('dd3dba4e-4baa-4bd3-95bf-64ffae9b7b51'
    /* '01/04/22', 'Skip making a request for resources that have already failed' */
    );
}
exports.isSkipKnownErroneousResourcesActivated = isSkipKnownErroneousResourcesActivated;
function isManualRedirectCancellationActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('3c6eb7e3-520b-41c6-b517-2f40a6dc2989'
    /* '08/16/2024',
        'bcleaver_ManuallyCancelRedirect'*/
    );
}
exports.isManualRedirectCancellationActivated = isManualRedirectCancellationActivated;
function isMsalTwoMinuteMaxRedirectKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ca6e64a5-3cde-4350-8a45-a21de378cc60'
    /* '07/14/22', 'Enable MSAL to hit max redirect limit in 2 minutes instead of 1' */
    );
}
exports.isMsalTwoMinuteMaxRedirectKsActivated = isMsalTwoMinuteMaxRedirectKsActivated;
function isMsalBrowserRetryAndSsoSilentImprovementsKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e0bd3586-89a3-466b-8ee0-db1f9457fe99'
    /* '07/14/22', 'Stop MSAL Browser from retrying endpoint resolution failures + increase timeout time' */
    );
}
exports.isMsalBrowserRetryAndSsoSilentImprovementsKsActivated = isMsalBrowserRetryAndSsoSilentImprovementsKsActivated;
function isMsalUseSeparateStorageKeysKsActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1a423106-78a1-4238-ba44-dd5f7f4138df'
    /* '08/18/22', 'Separate the storage parameters used for redirect flows by v1/v2' */
    );
}
exports.isMsalUseSeparateStorageKeysKsActivated = isMsalUseSeparateStorageKeysKsActivated;
function isTeamsBrowserHostedKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('9b9957d5-67d6-4f73-8882-11ee6bbaa066'
    /* '02/20/2024', 'Set teams hosted container' */
    );
}
exports.isTeamsBrowserHostedKSActive = isTeamsBrowserHostedKSActive;
function useOBOInWebForSecureBrokerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('043feae9-5a56-40ff-b175-6f7f813507ff' /* '10/26/22', 'use OBO in web for secure broker' */);
}
exports.useOBOInWebForSecureBrokerKSActivated = useOBOInWebForSecureBrokerKSActivated;
function isLogAfterFullPageRedirectKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5a350491-8d32-46a3-a4c4-fdce19a5f73e' /* '08/17/23', 'log after full page redirect' */);
}
exports.isLogAfterFullPageRedirectKSActivated = isLogAfterFullPageRedirectKSActivated;
function isMsalImplicitV1TokenByPopupKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e6246a29-f6af-41de-88b7-3f023102181b' /* '08/14/23', 'Enable acquiring a token by popup' */);
}
exports.isMsalImplicitV1TokenByPopupKSActivated = isMsalImplicitV1TokenByPopupKSActivated;
// cross project KS
function isAlwaysInitOboConfigKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('7fb9acc4-19f2-446f-b592-f15c49bfd771'
    /* '10/17/2023', 'Always initialize OBO configuration' */
    );
}
exports.isAlwaysInitOboConfigKSActive = isAlwaysInitOboConfigKSActive;
function isTenantAdminAppFirtPartyKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('43551625-af3d-44b0-8fb3-f0272435c606' /* '12/08/23', 'Is tenant admin app first party' */);
}
exports.isTenantAdminAppFirtPartyKSActivated = isTenantAdminAppFirtPartyKSActivated;
function isEnsureRefreshTokenKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('db350345-b6f8-4c00-9cbe-49ae316f6637' /* '12/06/2023', 'Disable ensure refresh token' */);
}
exports.isEnsureRefreshTokenKSActivated = isEnsureRefreshTokenKSActivated;
function isUpdateReplyUriForV1FallbackKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bd15098f-e57e-4171-8d51-c5c0c59dfaeb' /* '01/31/2024', 'isUpdateReplyUriForV1Fallback' */);
}
exports.isUpdateReplyUriForV1FallbackKSActivated = isUpdateReplyUriForV1FallbackKSActivated;
function isConditionalAccessIgnoredOnEnsureStateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('780272c0-bb66-4f68-a470-15c7c74493fd' /* '02/09/2024', 'isConditionalAccessIgnoredOnEnsureStateKSActivated' */);
}
exports.isConditionalAccessIgnoredOnEnsureStateKSActivated = isConditionalAccessIgnoredOnEnsureStateKSActivated;
function isLogAllFullPageRedirectsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('33e01c38-ae10-4a40-ae92-1d8f82214515' /* '04/24/2024', 'isLogAllFullPageRedirectsKSActivated' */);
}
exports.isLogAllFullPageRedirectsKSActivated = isLogAllFullPageRedirectsKSActivated;
function isFailAllAnonymousTokenRequestsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bba4380d-7882-4c9f-9033-0dba0b76ca93' /* '05/30/2024', 'isFailAllAnonymousTokenRequestsKSActivated' */);
}
exports.isFailAllAnonymousTokenRequestsKSActivated = isFailAllAnonymousTokenRequestsKSActivated;
function isFailAllEmailAuthGuestTokenRequestsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('52e15dc9-ebf3-41a7-af0f-ff9cf93ac3bd' /* '05/30/2024', 'isFailAllEmailAuthGuestTokenRequestsKSActivated' */);
}
exports.isFailAllEmailAuthGuestTokenRequestsKSActivated = isFailAllEmailAuthGuestTokenRequestsKSActivated;
function isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('748cee93-1dd8-4510-86b8-56446240c126' /* '05/13/2024', 'isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated' */);
}
exports.isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated = isOnlyEnsureServicePrincipalIdOnTokenFetchKSActivated;
function isDisableMsalV1KSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a079cff8-1a14-434c-811b-0e702c32de0c' /* '09/18/2024', 'isDisableMsalV1KSActivated' */);
}
exports.isDisableMsalV1KSActivated = isDisableMsalV1KSActivated;
function isEnablePopupFlowForMsalV3KSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('32a6db75-625e-451a-8813-6a88e019e71f' /* '07/01/2024', 'isPopupFlowEnabledForMsalV3' */);
}
exports.isEnablePopupFlowForMsalV3KSActivated = isEnablePopupFlowForMsalV3KSActivated;
function isEnableEncryptedAuthorizeResponseKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('0c1b5fc7-8d68-4ac0-8392-e2417c09a244' /* '09/10/2024', 'isEnableEncryptedAuthorizeResponseKSActivated' */);
}
exports.isEnableEncryptedAuthorizeResponseKSActivated = isEnableEncryptedAuthorizeResponseKSActivated;
function isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('57d6fc0d-c2df-4446-8a94-d9ccc9e8870b' /* '08/13/2024', 'isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated' */);
}
exports.isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated = isUseSessionStorageForRedirectScenarioProxyMonitoringKSActivated;
function isModifyPopupLogicToCallRequestPopupKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('28f74ae4-5d54-4766-91aa-ad95ed2ff2dd' /* '09/05/2024', 'isModifyPopupLogicToCallRequestPopupKSActivated' */);
}
exports.isModifyPopupLogicToCallRequestPopupKSActivated = isModifyPopupLogicToCallRequestPopupKSActivated;
function isHandleNAAInteractionRequiredWithPopupKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('11bf81db-a2df-4328-ae41-1d76788c5184' /* '09/12/2024', 'isHandleNAAInteractionRequiredWithPopupKSActivated' */);
}
exports.isHandleNAAInteractionRequiredWithPopupKSActivated = isHandleNAAInteractionRequiredWithPopupKSActivated;
function isLogCorrelationIdKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1020a2a1-2e9f-4e0f-8baa-bd7451c76ea1' /* '10/01/2024', 'isLogCorrelationIdKSActivated' */);
}
exports.isLogCorrelationIdKSActivated = isLogCorrelationIdKSActivated;
function isEnsureStateOnNAAKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b745926a-0c85-41a6-b4e2-42c7c14ee40a' /* '10/17/2024', 'isEnsureStateOnNAAKSActivated' */);
}
exports.isEnsureStateOnNAAKSActivated = isEnsureStateOnNAAKSActivated;
function isAlwaysUseMsalV3KSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('5b2d6b61-9c8c-44a7-8d95-cf0b833efee5' /* '10/04/2024', 'isAlwaysUseMsalV3KSActivated' */);
}
exports.isAlwaysUseMsalV3KSActivated = isAlwaysUseMsalV3KSActivated;
function isExpectedResultOnErrorReturnedKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a715e9eb-fd6f-4462-8404-06f4487407e2' /* '10/25/2024', 'isExpectedResultOnErrorReturnedKSActivated' */);
}
exports.isExpectedResultOnErrorReturnedKSActivated = isExpectedResultOnErrorReturnedKSActivated;
function isUseClientIndexKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('0cd7b4c0-2409-4e6c-8173-1d9516f9d8bc' /* '10/30/2024', 'isUseClientIndexKSActivated' */);
}
exports.isUseClientIndexKSActivated = isUseClientIndexKSActivated;
function isLogUserCancelledErrorAsExpectedFailureKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('58c93a33-ee2f-4900-98be-dde47c101e13' /* '11/14/2024', 'isLogUserCanceledErrorAsExpectedFailureKSActivated' */);
}
exports.isLogUserCancelledErrorAsExpectedFailureKSActivated = isLogUserCancelledErrorAsExpectedFailureKSActivated;
function isHandleClaimsChallengeOn1p3pExchangeFailureKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('62bd27b7-3caa-4bed-afb0-11ef71384035' /* '12/3/2024', 'isHandleClaimsChallengeOn1p3pExchangeFailureKSActivated' */);
}
exports.isHandleClaimsChallengeOn1p3pExchangeFailureKSActivated = isHandleClaimsChallengeOn1p3pExchangeFailureKSActivated;
function isUseUpdatedBrowserUtilitiesSPMobileCheckKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('370234a6-33bf-469b-bbcf-33df2d067a34' /* '1/5/2024', 'bentsai', 'isUseUpdatedBrowserUtilitiesSPMobileCheckKSActivated' */);
}
exports.isUseUpdatedBrowserUtilitiesSPMobileCheckKSActivated = isUseUpdatedBrowserUtilitiesSPMobileCheckKSActivated;
function isRetryMonitorWindowTimeoutKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('54bc6971-f3d4-4002-b69b-f1cab828126b' /* '01/22/2025', 'grahamc - retry monitor window timeout failures' */);
}
exports.isRetryMonitorWindowTimeoutKSActivated = isRetryMonitorWindowTimeoutKSActivated;
function isRemoveFloatingPromisesKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ec107b65-c663-467d-8436-f537185b214e' /* '01/24/2025', 'grahamc - remove floating promise when initializing MSAL' */);
}
exports.isRemoveFloatingPromisesKSActivated = isRemoveFloatingPromisesKSActivated;
function isUseUpdatedBrowserUtilitiesVCMCheckKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('78607411-c58f-4e2b-9144-b80c0d424be7' /* '2/4/2025', 'bentsai', 'isUseUpdatedBrowserUtilitiesVCMCheckKSActivated' */);
}
exports.isUseUpdatedBrowserUtilitiesVCMCheckKSActivated = isUseUpdatedBrowserUtilitiesVCMCheckKSActivated;
function isSkipCacheForSilentAuthKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6ed0c812-5a70-40d5-bf5d-326fbc30f491' /* '3/11/2025', 'brwatt' */);
}
exports.isSkipCacheForSilentAuthKSActivated = isSkipCacheForSilentAuthKSActivated;
