// @microsoft/sp-component-base/lib-commonjs/dataCollector/IWebPartLoadData.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports._CacheMissReason = exports.CacheStatus = void 0;
/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/typedef
exports.CacheStatus = {
    True: 'true',
    False: 'false',
    NotAvailable: 'N/A'
};
/**
 * @internal
 */
// eslint-disable-next-line @typescript-eslint/typedef
exports._CacheMissReason = {
    /**
     * Cache was available but config requested has changed.
     */
    CacheConfigMissmatch: 'CacheConfigMissmatch',
    /**
     * Cache did not arrvie by the time web part requested it.
     */
    LateFlush: 'LateFlush',
    /**
     * Cache is available but there is no cache entry for the web part.
     */
    CachedItemNotFound: 'CachedItemNotFound',
    /**
     * Cache is available, config is also good but is too stale for use.
     */
    CacheExpired: 'CacheExpired',
    /**
     * Cache miss happened but reason is not specified.
     */
    NotSpecified: 'NotSpecified',
    /**
     * Not Applicable if there was cache hit.
     */
    NotApplicable: 'N/A'
};
