// @microsoft/sp-property-pane/lib-commonjs/common/Flights.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isCardMotionEnabled = exports.isContentPanelRenderLoggingEnabled = exports.isPageDetailsEnabled = exports.isImagesODBFilePickerEnabled = exports.isModifyFontSizeTitleHeadingsEnabled = exports.isConfigurationToolsEnabled = exports.isContentPanelV2Enabled = exports.isContentPanelEnabled = exports.isCardDesignerPlusPlusFlight2Enabled = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
var KillSwitches_1 = require("./KillSwitches");
function isCardDesignerPlusPlusFlight2Enabled() {
    return sp_core_library_1._SPFlight.isEnabled(61147 /* CardDesignerPlusPlusPhase2Client */);
}
exports.isCardDesignerPlusPlusFlight2Enabled = isCardDesignerPlusPlusFlight2Enabled;
// cross-project flight
function isContentPanelEnabled() {
    return (sp_core_library_1._SPFlight.isEnabled(60808 /* ContentPanel */) ||
        sp_core_library_1._SPExperiment.getVariantAndLogExposure(90301 /* WEXPagesContentPanelExperiment */) === 1);
}
exports.isContentPanelEnabled = isContentPanelEnabled;
// cross-project flight
function isContentPanelV2Enabled() {
    return isContentPanelEnabled() && sp_core_library_1._SPFlight.isEnabled(62215 /* ContentPaneV2 */);
}
exports.isContentPanelV2Enabled = isContentPanelV2Enabled;
// cross-project flight
function isConfigurationToolsEnabled() {
    return isContentPanelEnabled() && !(0, KillSwitches_1.isConfigurationToolsKSActivated)();
}
exports.isConfigurationToolsEnabled = isConfigurationToolsEnabled;
// cross-project flight
function isModifyFontSizeTitleHeadingsEnabled() {
    return !(0, KillSwitches_1.isModifyFontSizeTitleHeadingKSctivated)() && sp_core_library_1._SPFlight.isEnabled(61245 /* WebPartTitleHeadings */);
}
exports.isModifyFontSizeTitleHeadingsEnabled = isModifyFontSizeTitleHeadingsEnabled;
// cross-project flight
function isImagesODBFilePickerEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(1342 /* ImagesODBFilePicker */);
}
exports.isImagesODBFilePickerEnabled = isImagesODBFilePickerEnabled;
// cross-project flight
function isPageDetailsEnabled() {
    return (isContentPanelEnabled() && sp_core_library_1._SPFlight.isEnabled(62215 /* ContentPaneV2 */) && !(0, KillSwitches_1.isPageDetailsKSActivated)());
}
exports.isPageDetailsEnabled = isPageDetailsEnabled;
//cross-project flight
function isContentPanelRenderLoggingEnabled() {
    return (sp_core_library_1._SPFlight.isEnabled(60808 /* ContentPanel */) && sp_core_library_1._SPFlight.isEnabled(62114 /* ContentPanelLogging */));
}
exports.isContentPanelRenderLoggingEnabled = isContentPanelRenderLoggingEnabled;
function isCardMotionEnabled() {
    return sp_core_library_1._SPFlight.isEnabled(62266 /* WexCardsMotion */);
}
exports.isCardMotionEnabled = isCardMotionEnabled;
