// @microsoft/sp-property-pane/lib-commonjs/common/KillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isReduceContentPanelControllerLogsKSActivated = exports.isAlignContentPanelControllerToggleAPIWithPropertyPaneKSActivated = exports.isFixNoTitleKSActivated = exports.isFixInlineLabelKSActivated = exports.isOverrideAriaLabelKSActivated = exports.isCollapseAndCustomizeDragImageKSActivated = exports.isMigrateToggleToSwitchKSActivated = exports.isPageDetailsKSActivated = exports.isSupport3rdPartyInContentPanelKSActivated = exports.isEnableHighlightItemExpandedStateKSActivated = exports.isLocalizeWebPartTitleHeadingLevelsKSActivated = exports.isAddTablistToProperyPaneKSActivated = exports.isSwitchToCommonScreenReaderAlertKSActivated = exports.isPropertyPaneTabsterEnabledKSActivated = exports.isAddUpdatedStylingClassForChoiceGroupKSActivated = exports.isAllowValidationForCustomFieldsKSActivated = exports.isAddTopMarginForExpandedPropertyPaneKSActivated = exports.isShowBackButtonBackKSActivated = exports.isTooltipForTextFieldEnabledKSActivated = exports.isA11yColorContrastChoiceGroupOptionKSActivated = exports.isAllowDisableThumbnailPickerKSActivated = exports.isConfigurationToolsKSActivated = exports.isModifyFontSizeTitleHeadingKSctivated = exports.isShowTemplateTextOverflowKSActivated = exports.isPropertyPaneGroupNameAndContentA11yKSActivated = exports.isApplyButtonFocusKSActivated = exports.isHideTitleBarKSActivated = exports.isImageChangeButtonFocusedKSActivated = exports.isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated = exports.isFluentIconPickerPerformanceKSActivated = exports.isForceCloseWithoutFiringEventsKSActivated = exports.isActivateActionOnRefreshKSActivated = exports.isHybridIconFontFamilyActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
// cross-project KS
function isHybridIconFontFamilyActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('65e35e80-07f0-4ed0-bb25-2cfdb4ab6bd5'
    /* '06/27/2022', 'Update MDL2 icon font family to the new hybrid font family to fix font not loading' */
    );
}
exports.isHybridIconFontFamilyActivated = isHybridIconFontFamilyActivated;
function isActivateActionOnRefreshKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8a933072-1f53-4ee9-b32f-95aced750f25'
    /* '02/07/2023', 'Refresh Action executed when property pane is not on screen' */
    );
}
exports.isActivateActionOnRefreshKSActivated = isActivateActionOnRefreshKSActivated;
function isForceCloseWithoutFiringEventsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6a65101f-91d5-462f-be36-f50e588af9c4'
    /* '07/01/2024', 'Changes force closing behavior to exclude firing events when switching between content pane and properties pane' */
    );
}
exports.isForceCloseWithoutFiringEventsKSActivated = isForceCloseWithoutFiringEventsKSActivated;
function isFluentIconPickerPerformanceKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4453188e-9e78-4dbe-b5ad-8dcbc492b797'
    /* '02/07/2023', 'Remove debouncer for performance upgrade on component' */
    );
}
exports.isFluentIconPickerPerformanceKSActivated = isFluentIconPickerPerformanceKSActivated;
function isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4453188e-9e78-4dbe-b5ad-8dcbc492b797'
    /* '04/07/2023', 'Different browsers generate different a11y tree. Disable ChoiceGroup aria-label and include it in radio button aria-label.' */
    );
}
exports.isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated = isIncludeChoiceGroupLabelWithRadioButtonAriaLabelKSActivated;
function isImageChangeButtonFocusedKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('86e8a66b-80f7-4f52-b085-9a3571dfc29c'
    /* '4/28/2023', 'Set focus on the change button after selecting an image from the file picker' */
    );
}
exports.isImageChangeButtonFocusedKSActivated = isImageChangeButtonFocusedKSActivated;
function isHideTitleBarKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ebeddd3a-683c-4846-9254-8fc217a57f0d'
    /* '5/16/2023', 'Add an option to hide the title bar' */
    );
}
exports.isHideTitleBarKSActivated = isHideTitleBarKSActivated;
function isApplyButtonFocusKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4b0e4e25-af2c-47b4-aab0-c6d6736ab370'
    /* '7/10/2023', 'set focus after clicking apply button' */
    );
}
exports.isApplyButtonFocusKSActivated = isApplyButtonFocusKSActivated;
function isPropertyPaneGroupNameAndContentA11yKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('66db35ed-35bd-4ef7-aa7b-7bea2c585abf'
    /* '8/25/2023', 'Add the relationship for the PropertyPane's Group name and content for fixing a11y issue' */
    );
}
exports.isPropertyPaneGroupNameAndContentA11yKSActivated = isPropertyPaneGroupNameAndContentA11yKSActivated;
function isShowTemplateTextOverflowKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('269bda49-b777-4f47-a207-de1fd3098165'
    /* '12/04/2023', 'Show template text overflow' */
    );
}
exports.isShowTemplateTextOverflowKSActivated = isShowTemplateTextOverflowKSActivated;
function isModifyFontSizeTitleHeadingKSctivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f517f548-0f32-4526-948d-f9422dd06fee'
    /* '12/08/2023', 'Modify size of title heading when level is changed' */
    );
}
exports.isModifyFontSizeTitleHeadingKSctivated = isModifyFontSizeTitleHeadingKSctivated;
// cross-project kill switch
function isConfigurationToolsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a3f45048-0db5-4b7c-9a9a-e3183d216c4d'
    /** '03/21/2024', 'Enables configuration tools pivot in content panel. */
    );
}
exports.isConfigurationToolsKSActivated = isConfigurationToolsKSActivated;
function isAllowDisableThumbnailPickerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('3f9bf326-f86b-4992-b5fe-87eae7ab8390'
    /* '12/14/2023', 'Allows thumbnail picker to be disabled' */
    );
}
exports.isAllowDisableThumbnailPickerKSActivated = isAllowDisableThumbnailPickerKSActivated;
function isA11yColorContrastChoiceGroupOptionKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('90d4af40-ef3c-47c9-a4ba-570de646d35b'
    /* '12/28/2023', 'Make light mode option text's color contrast accessible' */
    );
}
exports.isA11yColorContrastChoiceGroupOptionKSActivated = isA11yColorContrastChoiceGroupOptionKSActivated;
function isTooltipForTextFieldEnabledKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('89d42863-7bb6-40ff-8688-4d97aa42fce0'
    /* '02/05/2024', 'Add tooltip for TextField' */
    );
}
exports.isTooltipForTextFieldEnabledKSActivated = isTooltipForTextFieldEnabledKSActivated;
function isShowBackButtonBackKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('81d64493-e421-402c-893b-87aaab9de576'
    /* '03/19/2024', 'Show back icon in case of nested header' */
    );
}
exports.isShowBackButtonBackKSActivated = isShowBackButtonBackKSActivated;
function isAddTopMarginForExpandedPropertyPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4283af84-d6f3-4cd9-ad86-d23e2848fb4b'
    /* '03/19/2024', 'Add top margin for expanded property pane content' */
    );
}
exports.isAddTopMarginForExpandedPropertyPaneKSActivated = isAddTopMarginForExpandedPropertyPaneKSActivated;
function isAllowValidationForCustomFieldsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ba80e564-a99d-4090-bacd-8c54c4f58d64'
    /* '03/22/2024', 'Allow validation for custom fields' */
    );
}
exports.isAllowValidationForCustomFieldsKSActivated = isAllowValidationForCustomFieldsKSActivated;
function isAddUpdatedStylingClassForChoiceGroupKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('fc584b86-b85c-47bb-8c94-3c5539acb4c5'
    /* '03/29/2024', 'Add styling class for choice group' */
    );
}
exports.isAddUpdatedStylingClassForChoiceGroupKSActivated = isAddUpdatedStylingClassForChoiceGroupKSActivated;
function isPropertyPaneTabsterEnabledKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c7dc4f94-8515-4aa7-9360-b1c5db92f079'
    /* '04/39/2024', 'Enable property pane tabster' */
    );
}
exports.isPropertyPaneTabsterEnabledKSActivated = isPropertyPaneTabsterEnabledKSActivated;
function isSwitchToCommonScreenReaderAlertKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('97f81811-a534-48f6-b3e1-a0f0268ac1ab' /* '05/20/2024', 'Switch to common screen reader alert' */);
}
exports.isSwitchToCommonScreenReaderAlertKSActivated = isSwitchToCommonScreenReaderAlertKSActivated;
function isAddTablistToProperyPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('0743240a-8e58-40ca-9d1c-6e69774f8a20' /* '11/04/2024', 'Add tablist to Property Pane' */);
}
exports.isAddTablistToProperyPaneKSActivated = isAddTablistToProperyPaneKSActivated;
function isLocalizeWebPartTitleHeadingLevelsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('7f699815-7ab3-4cc0-be35-e6b71efdd983' /* '12/27/2024', 'Localizes the web part title heading level dropdown option labels.' */);
}
exports.isLocalizeWebPartTitleHeadingLevelsKSActivated = isLocalizeWebPartTitleHeadingLevelsKSActivated;
// cross-project KS
function isEnableHighlightItemExpandedStateKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8d7b5df9-8690-4a72-9cc3-70e7f595c8fc'
    /* '1/13/2025', ' (dreamli) Enable the index and expanded state of the highlightItem to be controlled separately.' */
    );
}
exports.isEnableHighlightItemExpandedStateKSActivated = isEnableHighlightItemExpandedStateKSActivated;
function isSupport3rdPartyInContentPanelKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('0743240a-8e58-40ca-9d1c-6e69774f8a20' /* '01/15/2025', 'Adds support for 3rd party users in content panel' */);
}
exports.isSupport3rdPartyInContentPanelKSActivated = isSupport3rdPartyInContentPanelKSActivated;
// cross-project kill switch
function isPageDetailsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('7518a1e3-4b70-474b-8030-dfa34e60a480'
    /** '01/22/2025', 'mparrachavez- Enables page details pivot in content panel. */
    );
}
exports.isPageDetailsKSActivated = isPageDetailsKSActivated;
function isMigrateToggleToSwitchKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('379a3743-d608-4f61-9508-da931c795441' /* '02/19/2025', 'ctanlimco - migrate fluentui toggle to switch' */);
}
exports.isMigrateToggleToSwitchKSActivated = isMigrateToggleToSwitchKSActivated;
function isCollapseAndCustomizeDragImageKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bc9118ed-6dc1-454a-a9c6-3562faa628bb'
    /* '1/20/2025', ' (dreamli) Collapse item and customize drag image when dragging the items in sortable list.' */
    );
}
exports.isCollapseAndCustomizeDragImageKSActivated = isCollapseAndCustomizeDragImageKSActivated;
function isOverrideAriaLabelKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('33875d58-ed85-478a-ac97-ef412fdb5998'
    /* '01/24/2025', 'ctanlimco - override aria label if toggle label tooltip is present' */
    );
}
exports.isOverrideAriaLabelKSActivated = isOverrideAriaLabelKSActivated;
function isFixInlineLabelKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ddeb57e6-703b-4943-8fe4-b53a6b673d96'
    /* '02/19/2025', 'ctanlimco - fix inline label with Switch' */
    );
}
exports.isFixInlineLabelKSActivated = isFixInlineLabelKSActivated;
function isFixNoTitleKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e8d24b23-4f99-499a-b8e0-3a940d605e0b'
    /* '02/19/2025', 'ctanlimco - fix no title provided for Switch */
    );
}
exports.isFixNoTitleKSActivated = isFixNoTitleKSActivated;
function isAlignContentPanelControllerToggleAPIWithPropertyPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('afda1d5e-1f3a-48bd-92a9-ee71fd14dcb0'
    /* '02/28/2025', 'jayharris - Aligns the toggle api in ContentPanelController with PropertyPane */
    );
}
exports.isAlignContentPanelControllerToggleAPIWithPropertyPaneKSActivated = isAlignContentPanelControllerToggleAPIWithPropertyPaneKSActivated;
function isReduceContentPanelControllerLogsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('33e8d405-9334-4fd9-91d4-6d23130800a8'
    /* '02/26/2025', 'jayharris - skips registering consumer if already registered */
    );
}
exports.isReduceContentPanelControllerLogsKSActivated = isReduceContentPanelControllerLogsKSActivated;
