// @microsoft/rush-lib/lib-esnext/pluginFramework/RushLifeCycle.js
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
import { AsyncParallelHook, AsyncSeriesHook, HookMap } from 'tapable';
/**
 * Hooks into the lifecycle of the Rush process invocation that plugins may tap into.
 *
 * @beta
 */
export class RushLifecycleHooks {
    constructor() {
        /**
         * The hook to run before executing any Rush CLI Command.
         */
        this.initialize = new AsyncSeriesHook(['command'], 'initialize');
        /**
         * The hook to run before executing any global Rush CLI Command (defined in command-line.json).
         */
        this.runAnyGlobalCustomCommand = new AsyncSeriesHook(['command'], 'runAnyGlobalCustomCommand');
        /**
         * A hook map to allow plugins to hook specific named global commands (defined in command-line.json) before execution.
         */
        this.runGlobalCustomCommand = new HookMap((key) => {
            return new AsyncSeriesHook(['command'], key);
        }, 'runGlobalCustomCommand');
        /**
         * The hook to run before executing any phased Rush CLI Command (defined in command-line.json, or the default "build" or "rebuild").
         */
        this.runAnyPhasedCommand = new AsyncSeriesHook(['command'], 'runAnyPhasedCommand');
        /**
         * A hook map to allow plugins to hook specific named phased commands (defined in command-line.json) before execution.
         */
        this.runPhasedCommand = new HookMap((key) => {
            return new AsyncSeriesHook(['command'], key);
        }, 'runPhasedCommand');
        /**
         * The hook to run between preparing the common/temp folder and invoking the package manager during "rush install" or "rush update".
         */
        this.beforeInstall = new AsyncSeriesHook(['command', 'subspace', 'variant'], 'beforeInstall');
        /**
         * The hook to run after a successful install.
         */
        this.afterInstall = new AsyncSeriesHook(['command', 'subspace', 'variant'], 'afterInstall');
        /**
         * A hook to allow plugins to hook custom logic to process telemetry data.
         */
        this.flushTelemetry = new AsyncParallelHook(['telemetryData'], 'flushTelemetry');
    }
}
