// @microsoft/rush-lib/lib-commonjs/logic/base/BaseProjectShrinkwrapFile.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.BaseProjectShrinkwrapFile = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
const RushConstants_1 = require("../RushConstants");
/**
 * This class handles creating the project/.rush/temp/shrinkwrap-deps.json file
 * which tracks the direct and indirect dependencies that a project consumes. This is used
 * to better determine which projects should be rebuilt when dependencies are updated.
 */
class BaseProjectShrinkwrapFile {
    constructor(shrinkwrapFile, project) {
        this.project = project;
        this.projectShrinkwrapFilePath = BaseProjectShrinkwrapFile.getFilePathForProject(this.project);
        this.shrinkwrapFile = shrinkwrapFile;
    }
    /**
     * Save an empty project shrinkwrap file. This is used in repos with no dependencies.
     */
    static async saveEmptyProjectShrinkwrapFileAsync(project) {
        const projectShrinkwrapFilePath = BaseProjectShrinkwrapFile.getFilePathForProject(project);
        await node_core_library_1.JsonFile.saveAsync({}, projectShrinkwrapFilePath, { ensureFolderExists: true });
    }
    /**
     * Get the fully-qualified path to the <project>/.rush/temp/shrinkwrap-deps.json
     * for the specified project.
     */
    static getFilePathForProject(project) {
        return `${project.projectRushTempFolder}/${RushConstants_1.RushConstants.projectShrinkwrapFilename}`;
    }
    /**
     * If the <project>/.rush/temp/shrinkwrap-deps.json file exists, delete it. Otherwise, do nothing.
     */
    async deleteIfExistsAsync() {
        await node_core_library_1.FileSystem.deleteFileAsync(this.projectShrinkwrapFilePath, { throwIfNotExists: false });
    }
}
exports.BaseProjectShrinkwrapFile = BaseProjectShrinkwrapFile;
