// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/FilePicker.module.scss.js
require("./FilePicker.module.css");
const styles = {
    filePicker: 'filePicker_20c8f110',
    tabsContainer: 'tabsContainer_20c8f110',
    tabContainer: 'tabContainer_20c8f110',
    nav: 'nav_20c8f110',
    tabHeaderContainer: 'tabHeaderContainer_20c8f110',
    breadcrumbNav: 'breadcrumbNav_20c8f110',
    breadcrumbNavItem: 'breadcrumbNavItem_20c8f110',
    tab: 'tab_20c8f110',
    tabOffset: 'tabOffset_20c8f110',
    tabFiles: 'tabFiles_20c8f110',
    scrollablePaneWrapper: 'scrollablePaneWrapper_20c8f110',
    actionButtonsContainer: 'actionButtonsContainer_20c8f110',
    actionButtons: 'actionButtons_20c8f110',
    actionButton: 'actionButton_20c8f110',
    tabHeader: 'tabHeader_20c8f110',
    selectedTab: 'selectedTab_20c8f110',
    header: 'header_20c8f110',
    xlg: 'xlg_20c8f110',
    xxlg: 'xxlg_20c8f110',
    xxxlg: 'xxxlg_20c8f110',
    xxxxlg: 'xxxxlg_20c8f110',
    itemPickerTopBar: 'itemPickerTopBar_20c8f110',
    commandBarNoChevron: 'commandBarNoChevron_20c8f110',
    fileItem: 'fileItem_20c8f110',
    folderItem: 'folderItem_20c8f110',
    fileTypeIcon: 'fileTypeIcon_20c8f110',
    fileTypeIconIcon: 'fileTypeIconIcon_20c8f110',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_20c8f110',
    iconColumnHeader: 'iconColumnHeader_20c8f110',
    fileRow: 'fileRow_20c8f110',
    folderRow: 'folderRow_20c8f110',
    emptyFolder: 'emptyFolder_20c8f110',
    emptyFolderImage: 'emptyFolderImage_20c8f110',
    emptyFolderImageTag: 'emptyFolderImageTag_20c8f110',
    emptyFolderTitle: 'emptyFolderTitle_20c8f110',
    emptyFolderSubText: 'emptyFolderSubText_20c8f110',
    emptyFolderMobile: 'emptyFolderMobile_20c8f110',
    emptyFolderPc: 'emptyFolderPc_20c8f110',
    emptyFolderIcon: 'emptyFolderIcon_20c8f110'
};
export default styles;
