// @microsoft/spfx-heft-plugins/lib-commonjs/plugins/deployAzureStorage/DeployAzureStoragePlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return DeployAzureStoragePlugin;
    }
});
const _fastglob = /*#__PURE__*/ _interop_require_default(require("fast-glob"));
const _heftconfigfile = require("@rushstack/heft-config-file");
const _terminal = require("@rushstack/terminal");
const _PluginNames = require("../PluginNames");
const _AzureUploader = require("../../utilities/AzureUploader");
const _devDeployUtilities = require("../../utilities/devDeployUtilities");
const _deployazurestorageschemajson = /*#__PURE__*/ _interop_require_default(require("./deploy-azure-storage.schema.json"));
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
const configFile = new _heftconfigfile.ConfigurationFile({
    projectRelativeFilePath: 'config/deploy-azure-storage.json',
    jsonSchemaObject: _deployazurestorageschemajson.default
});
class DeployAzureStoragePlugin {
    static async deployToAzureAsync({ terminal, buildFolder, emitError, rigConfig, tryCreateContainer }) {
        const config = await configFile.tryLoadConfigurationFileForProjectAsync(terminal, buildFolder, rigConfig) || {};
        if (!config.workingDir) {
            config.workingDir = `${buildFolder}/release/assets`;
        } else {
            // Trim trailing slashes
            config.workingDir = config.workingDir.replace(/[\/\\]+$/, '');
        }
        if (!config.uploadPath) {
            config.uploadPath = '**/*';
        }
        if (!config.container) {
            emitError(new Error('Config file missing container name.'));
            return;
        }
        if (!config.account) {
            emitError(new Error('Config file missing Azure account name.'));
            return;
        }
        if (!config.accessKey) {
            emitError(new Error('Config file missing Azure access key.'));
            return;
        }
        terminal.writeLine(`Uploading files '${config.uploadPath}' from directory '${config.workingDir}' to Azure`);
        const files = await (0, _fastglob.default)(config.uploadPath, {
            cwd: config.workingDir
        });
        const filesToUpload = files.map((filename)=>{
            return {
                azurePath: filename,
                localPath: `${config.workingDir}/${filename}`
            };
        });
        const azureUploader = new _AzureUploader.AzureUploader({
            storageAccountName: config.account,
            containerName: config.container,
            storageKey: config.accessKey
        });
        try {
            if (tryCreateContainer) {
                await azureUploader.ensureContainerExistsAsync(terminal);
            }
            await azureUploader.uploadFilesToAzureAsync(terminal, filesToUpload);
        } catch (e) {
            emitError(e);
        }
        terminal.writeLine(_terminal.Colorize.green('Upload complete!\n'));
        terminal.writeLine(`Access your files at: ${(0, _devDeployUtilities.getStorageAccountUrl)(config.account)}/${config.container}\n`);
    }
    apply(taskSession, heftConfiguration) {
        const scopedLogger = taskSession.logger;
        const skipTryCreateContainerParameter = taskSession.parameters.getFlagParameter('--skip-try-create-container');
        // TODO: Add incremental upload support
        taskSession.hooks.run.tapPromise(_PluginNames.DeployAzureStoragePlugin, async ()=>{
            await DeployAzureStoragePlugin.deployToAzureAsync({
                terminal: heftConfiguration.globalTerminal,
                emitError: scopedLogger.emitError.bind(scopedLogger),
                buildFolder: heftConfiguration.buildFolderPath,
                // TODO: Update this to use an interface instead of a class
                rigConfig: heftConfiguration.rigConfig,
                tryCreateContainer: !skipTryCreateContainerParameter.value
            });
        });
    }
}

//#sourceMappingUrl=./DeployAzureStoragePlugin.js.map
