// @microsoft/rush-lib/lib-commonjs/logic/CredentialCache.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CredentialCache = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
const Utilities_1 = require("../utilities/Utilities");
const RushUserConfiguration_1 = require("../api/RushUserConfiguration");
const credentials_schema_json_1 = __importDefault(require("../schemas/credentials.schema.json"));
const objectUtilities_1 = require("../utilities/objectUtilities");
const CACHE_FILENAME = 'credentials.json';
const LATEST_CREDENTIALS_JSON_VERSION = '0.1.0';
/**
 * @beta
 */
class CredentialCache /* implements IDisposable */ {
    constructor(cacheFilePath, loadedJson, lockfile) {
        this._modified = false;
        this._disposed = false;
        if (loadedJson && loadedJson.version !== LATEST_CREDENTIALS_JSON_VERSION) {
            throw new Error(`Unexpected credentials.json file version: ${loadedJson.version}`);
        }
        this._cacheFilePath = cacheFilePath;
        this._cacheEntries = new Map(Object.entries((loadedJson === null || loadedJson === void 0 ? void 0 : loadedJson.cacheEntries) || {}));
        this._supportsEditing = !!lockfile;
        this._lockfile = lockfile;
    }
    static async initializeAsync(options) {
        const rushUserFolderPath = RushUserConfiguration_1.RushUserConfiguration.getRushUserFolderPath();
        const cacheFilePath = `${rushUserFolderPath}/${CACHE_FILENAME}`;
        const jsonSchema = node_core_library_1.JsonSchema.fromLoadedObject(credentials_schema_json_1.default);
        let loadedJson;
        try {
            loadedJson = await node_core_library_1.JsonFile.loadAndValidateAsync(cacheFilePath, jsonSchema);
        }
        catch (e) {
            if (!node_core_library_1.FileSystem.isErrnoException(e)) {
                throw e;
            }
        }
        let lockfile;
        if (options.supportEditing) {
            lockfile = await node_core_library_1.LockFile.acquireAsync(rushUserFolderPath, `${CACHE_FILENAME}.lock`);
        }
        const credentialCache = new CredentialCache(cacheFilePath, loadedJson, lockfile);
        return credentialCache;
    }
    static async usingAsync(options, doActionAsync) {
        await Utilities_1.Utilities.usingAsync(async () => await CredentialCache.initializeAsync(options), doActionAsync);
    }
    setCacheEntry(cacheId, entry) {
        this._validate(true);
        const { expires, credential, credentialMetadata } = entry;
        const expiresMilliseconds = (expires === null || expires === void 0 ? void 0 : expires.getTime()) || 0;
        const existingCacheEntry = this._cacheEntries.get(cacheId);
        if ((existingCacheEntry === null || existingCacheEntry === void 0 ? void 0 : existingCacheEntry.credential) !== credential ||
            (existingCacheEntry === null || existingCacheEntry === void 0 ? void 0 : existingCacheEntry.expires) !== expiresMilliseconds ||
            !(0, objectUtilities_1.objectsAreDeepEqual)(existingCacheEntry === null || existingCacheEntry === void 0 ? void 0 : existingCacheEntry.credentialMetadata, credentialMetadata)) {
            this._modified = true;
            this._cacheEntries.set(cacheId, {
                expires: expiresMilliseconds,
                credential,
                credentialMetadata
            });
        }
    }
    tryGetCacheEntry(cacheId) {
        this._validate(false);
        const cacheEntry = this._cacheEntries.get(cacheId);
        if (cacheEntry) {
            const result = {
                expires: cacheEntry.expires ? new Date(cacheEntry.expires) : undefined,
                credential: cacheEntry.credential,
                credentialMetadata: cacheEntry.credentialMetadata
            };
            return result;
        }
        else {
            return undefined;
        }
    }
    deleteCacheEntry(cacheId) {
        this._validate(true);
        if (this._cacheEntries.has(cacheId)) {
            this._modified = true;
            this._cacheEntries.delete(cacheId);
        }
    }
    trimExpiredEntries() {
        this._validate(true);
        const now = Date.now();
        for (const [cacheId, cacheEntry] of this._cacheEntries.entries()) {
            if (cacheEntry.expires < now) {
                this._cacheEntries.delete(cacheId);
                this._modified = true;
            }
        }
    }
    async saveIfModifiedAsync() {
        this._validate(true);
        if (this._modified) {
            const cacheEntriesJson = {};
            for (const [cacheId, cacheEntry] of this._cacheEntries.entries()) {
                cacheEntriesJson[cacheId] = cacheEntry;
            }
            const newJson = {
                version: LATEST_CREDENTIALS_JSON_VERSION,
                cacheEntries: cacheEntriesJson
            };
            await node_core_library_1.JsonFile.saveAsync(newJson, this._cacheFilePath, {
                ensureFolderExists: true,
                updateExistingFile: true,
                ignoreUndefinedValues: true
            });
            this._modified = false;
        }
    }
    dispose() {
        var _a;
        (_a = this._lockfile) === null || _a === void 0 ? void 0 : _a.release();
        this._disposed = true;
    }
    _validate(requiresEditing) {
        if (!this._supportsEditing && requiresEditing) {
            throw new Error(`This instance of ${CredentialCache.name} does not support editing.`);
        }
        if (this._disposed) {
            throw new Error(`This instance of ${CredentialCache.name} has been disposed.`);
        }
    }
}
exports.CredentialCache = CredentialCache;
