// @microsoft/rush-lib/lib-esnext/logic/versionMismatch/VersionMismatchFinderProject.js
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
import { VersionMismatchFinderEntity } from './VersionMismatchFinderEntity';
export class VersionMismatchFinderProject extends VersionMismatchFinderEntity {
    constructor(project) {
        super({
            friendlyName: project.packageName,
            decoupledLocalDependencies: project.decoupledLocalDependencies,
            skipRushCheck: project.skipRushCheck
        });
        this._fileManager = project.packageJsonEditor;
        this.packageName = project.packageName;
    }
    get filePath() {
        return this._fileManager.filePath;
    }
    get allDependencies() {
        return [...this._fileManager.dependencyList, ...this._fileManager.devDependencyList];
    }
    tryGetDependency(packageName) {
        return this._fileManager.tryGetDependency(packageName);
    }
    tryGetDevDependency(packageName) {
        return this._fileManager.tryGetDevDependency(packageName);
    }
    addOrUpdateDependency(packageName, newVersion, dependencyType) {
        return this._fileManager.addOrUpdateDependency(packageName, newVersion, dependencyType);
    }
    removeDependency(packageName, dependencyType) {
        return this._fileManager.removeDependency(packageName, dependencyType);
    }
    saveIfModified() {
        return this._fileManager.saveIfModified();
    }
}
