// @microsoft/sp-build-core-tasks/lib-commonjs/packageSolution/PackageSolutionTask.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "PackageSolutionTask", {
    enumerable: true,
    get: function() {
        return PackageSolutionTask;
    }
});
const _terminal = require("@rushstack/terminal");
const _spfxheftplugins = require("@microsoft/spfx-heft-plugins");
const _OdspGulpTask = require("../OdspGulpTask");
const _GulpTaskTerminalProvider = require("../utilities/GulpTaskTerminalProvider");
class PackageSolutionTask extends _OdspGulpTask.OdspGulpTask {
    async executeTask() {
        const solutionPackager = new _spfxheftplugins._SolutionPackager({
            buildFolder: this.buildConfig.rootPath,
            terminal: new _terminal.Terminal(new _GulpTaskTerminalProvider.GulpTaskTerminalProvider(this)),
            emitWarning: (error)=>this.logWarning(error.message),
            emitError: (error)=>this.logError(error.message)
        });
        await solutionPackager.packageSolutionAsync(this.buildConfig.production);
    }
    constructor(){
        super('package-solution');
    }
}

//#sourceMappingUrl=./PackageSolutionTask.js.map
