// @microsoft/rush-lib/lib-commonjs/api/ApprovedPackagesPolicy.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ApprovedPackagesPolicy = void 0;
const path = __importStar(require("path"));
const ApprovedPackagesConfiguration_1 = require("./ApprovedPackagesConfiguration");
const RushConstants_1 = require("../logic/RushConstants");
/**
 * This is a helper object for RushConfiguration.
 * It exposes the "approvedPackagesPolicy" feature from rush.json.
 * @public
 */
class ApprovedPackagesPolicy {
    /** @internal */
    constructor(rushConfiguration, rushConfigurationJson) {
        const approvedPackagesPolicy = rushConfigurationJson.approvedPackagesPolicy || {};
        this.enabled = !!rushConfigurationJson.approvedPackagesPolicy;
        this.ignoredNpmScopes = new Set(approvedPackagesPolicy.ignoredNpmScopes);
        this.reviewCategories = new Set(approvedPackagesPolicy.reviewCategories);
        if (this.enabled) {
            if (!this.reviewCategories.size) {
                throw new Error(`The "approvedPackagesPolicy" feature is enabled ${RushConstants_1.RushConstants.rushJsonFilename}, but the reviewCategories` +
                    ` list is not configured.`);
            }
        }
        // Load browser-approved-packages.json
        const browserApprovedPackagesPath = path.join(rushConfiguration.commonRushConfigFolder, RushConstants_1.RushConstants.browserApprovedPackagesFilename);
        this.browserApprovedPackages = new ApprovedPackagesConfiguration_1.ApprovedPackagesConfiguration(browserApprovedPackagesPath);
        this.browserApprovedPackages.tryLoadFromFile(this.enabled);
        // Load nonbrowser-approved-packages.json
        const nonbrowserApprovedPackagesPath = path.join(rushConfiguration.commonRushConfigFolder, RushConstants_1.RushConstants.nonbrowserApprovedPackagesFilename);
        this.nonbrowserApprovedPackages = new ApprovedPackagesConfiguration_1.ApprovedPackagesConfiguration(nonbrowserApprovedPackagesPath);
        this.nonbrowserApprovedPackages.tryLoadFromFile(this.enabled);
    }
}
exports.ApprovedPackagesPolicy = ApprovedPackagesPolicy;
