// @microsoft/sp-property-pane/lib-commonjs/propertyPaneTablist/PropertyPaneTabList.styles.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPropertyPaneClassNames = exports.getPropertyPaneStyles = void 0;
var react_components_1 = require("@fluentui/react-components");
var office_ui_fabric_react_bundle_1 = require("@ms/office-ui-fabric-react-bundle");
var getPropertyPaneStyles = function () {
    return {
        tablist: {
            borderBottom: '1px solid',
            borderBottomColor: react_components_1.tokens.colorNeutralStroke2,
            marginBottom: '6px',
            paddingLeft: '10px',
            paddingTop: '2px'
        },
        tabItem: {
            paddingTop: '8px'
        },
        menu: {
            backgroundColor: react_components_1.tokens.colorNeutralBackground1
        },
        menuButton: {
            alignSelf: 'center'
        },
        horizontalTablist: {
            overflow: 'hidden',
            padding: '5px',
            zIndex: 0,
            height: 'fit-content'
        }
    };
};
exports.getPropertyPaneStyles = getPropertyPaneStyles;
exports.getPropertyPaneClassNames = (0, office_ui_fabric_react_bundle_1.classNamesFunction)();
