// @microsoft/spfx-heft-plugins/lib-esm/plugins/packageSolutionPlugin/packageSolution/xmlGenerators/createFeatureXml.js
/**
 * Creates an package's Feature.Xml
 */ import * as path from 'path';
import * as uuid from 'uuid';
import XmlGenerator from 'xml';
import { merge } from 'lodash';
import { FileSystem } from '@rushstack/node-core-library';
import createRelationshipsXml from './createRelationshipsXml';
import createComponentXml from './createComponentXml';
import { readSharepointAssets } from '../utils/readSharepointAssets';
export default function createFeatureXml(terminal, xmlSerializationOptions, feature, resources, clientSideAssets, featureXmlFile = `feature_${feature.id}.xml`) {
    feature = merge({
        version: '1.0.0.0'
    }, feature);
    const featureXml = {
        filename: featureXmlFile,
        // eslint-disable-next-line @rushstack/no-null
        relationships: null,
        contents: Buffer.from(XmlGenerator({
            Feature: [
                {
                    _attr: {
                        xmlns: 'http://schemas.microsoft.com/sharepoint/',
                        Title: feature.title,
                        Description: feature.description,
                        Id: feature.id,
                        Version: feature.version,
                        Scope: 'Web',
                        Hidden: 'FALSE'
                    }
                }
            ]
        }, xmlSerializationOptions))
    };
    const configXml = {
        filename: `${featureXmlFile}.config.xml`,
        contents: Buffer.from(XmlGenerator({
            AppPartConfig: [
                {
                    _attr: {
                        xmlns: 'http://schemas.microsoft.com/sharepoint/2012/app/partconfiguration'
                    }
                },
                {
                    Id: uuid.v4()
                }
            ]
        }, xmlSerializationOptions))
    };
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const componentsXml = feature.components.map((component)=>{
        return createComponentXml(xmlSerializationOptions, component, feature.id);
    });
    const clientSideResources = [];
    if (clientSideAssets && clientSideAssets.length) {
        // read each hosted asset
        // iterate through every file that was found and add a relationship
        clientSideAssets.forEach(({ originalFilename, packageFilename })=>{
            // read the file into memory
            const contents = FileSystem.readFileToBuffer(originalFilename);
            const relativeFilename = path.posix.join('ClientSideAssets', packageFilename);
            const file = {
                contents,
                filename: relativeFilename
            };
            clientSideResources.push(file);
        });
    }
    // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
    const assets = readSharepointAssets(terminal, feature.assets, feature.id);
    const relationshipsXml = createRelationshipsXml(xmlSerializationOptions, featureXmlFile, [
        {
            type: 'partconfiguration',
            target: configXml.filename
        }
    ].concat(componentsXml.map((component)=>{
        return {
            type: 'feature-elementmanifest',
            target: component.filename
        };
    })).concat(resources.localizedResourcePaths.length > 0 ? resources.localizedResourcePaths.map((filename)=>{
        return {
            type: 'content-resource',
            target: filename
        };
    }) : []).concat(resources.defaultResourcePath ? {
        type: 'content-defaultresource',
        target: resources.defaultResourcePath
    } : []).concat(createAssetRelationships(assets)).concat(clientSideResources.map((clientSideResource)=>{
        return {
            target: clientSideResource.filename,
            type: 'clientsideasset'
        };
    })));
    return {
        filename: featureXml.filename,
        contents: featureXml.contents,
        relationships: relationshipsXml,
        config: configXml,
        components: componentsXml,
        assets,
        clientSideResources
    };
}
function createAssetRelationships(assets) {
    let relationships = [];
    if (assets) {
        relationships = relationships.concat(createAssetRelationship('feature-elementfile', assets.elementFiles), createAssetRelationship('feature-elementmanifest', assets.elementManifests), createAssetRelationship('feature-upgradeactions', assets.upgradeActions));
    }
    return relationships;
}
function createAssetRelationship(type, files = []) {
    return files.map((file)=>{
        return {
            type,
            target: file.filename
        };
    });
}

//#sourceMappingUrl=./createFeatureXml.js.map
