// @microsoft/rush-lib/lib-esnext/api/Variants.js
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
import { EnvironmentVariableNames } from './EnvironmentConfiguration';
import { RushConstants } from '../logic/RushConstants';
/**
 * Provides the parameter configuration for '--variant'.
 */
export const VARIANT_PARAMETER = {
    parameterLongName: '--variant',
    argumentName: 'VARIANT',
    description: 'Run command using a variant installation configuration',
    environmentVariable: EnvironmentVariableNames.RUSH_VARIANT
};
export async function getVariantAsync(variantsParameter, rushConfiguration, defaultToCurrentlyInstalledVariant) {
    let variant = variantsParameter === null || variantsParameter === void 0 ? void 0 : variantsParameter.value;
    if (variant && !rushConfiguration.variants.has(variant)) {
        throw new Error(`The variant "${variant}" is not defined in ${RushConstants.rushJsonFilename}`);
    }
    if (!variant && defaultToCurrentlyInstalledVariant) {
        variant = await rushConfiguration.getCurrentlyInstalledVariantAsync();
    }
    return variant;
}
