// @microsoft/rush-lib/lib-commonjs/cli/scriptActions/PhasedScriptAction.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhasedScriptAction = void 0;
const node_core_library_1 = require("@rushstack/node-core-library");
const terminal_1 = require("@rushstack/terminal");
const PhasedCommandHooks_1 = require("../../pluginFramework/PhasedCommandHooks");
const SetupChecks_1 = require("../../logic/SetupChecks");
const Stopwatch_1 = require("../../utilities/Stopwatch");
const BaseScriptAction_1 = require("./BaseScriptAction");
const OperationExecutionManager_1 = require("../../logic/operations/OperationExecutionManager");
const RushConstants_1 = require("../../logic/RushConstants");
const EnvironmentConfiguration_1 = require("../../api/EnvironmentConfiguration");
const BuildCacheConfiguration_1 = require("../../api/BuildCacheConfiguration");
const SelectionParameterSet_1 = require("../parsing/SelectionParameterSet");
const PhasedOperationPlugin_1 = require("../../logic/operations/PhasedOperationPlugin");
const ShellOperationRunnerPlugin_1 = require("../../logic/operations/ShellOperationRunnerPlugin");
const EventHooks_1 = require("../../api/EventHooks");
const ProjectChangeAnalyzer_1 = require("../../logic/ProjectChangeAnalyzer");
const OperationStatus_1 = require("../../logic/operations/OperationStatus");
const OperationResultSummarizerPlugin_1 = require("../../logic/operations/OperationResultSummarizerPlugin");
const ParseParallelism_1 = require("../parsing/ParseParallelism");
const CobuildConfiguration_1 = require("../../api/CobuildConfiguration");
const CacheableOperationPlugin_1 = require("../../logic/operations/CacheableOperationPlugin");
const RushProjectConfiguration_1 = require("../../api/RushProjectConfiguration");
const LegacySkipPlugin_1 = require("../../logic/operations/LegacySkipPlugin");
const ValidateOperationsPlugin_1 = require("../../logic/operations/ValidateOperationsPlugin");
const ShardedPhaseOperationPlugin_1 = require("../../logic/operations/ShardedPhaseOperationPlugin");
const FlagFile_1 = require("../../api/FlagFile");
const WeightedOperationPlugin_1 = require("../../logic/operations/WeightedOperationPlugin");
const Variants_1 = require("../../api/Variants");
const Selection_1 = require("../../logic/Selection");
const NodeDiagnosticDirPlugin_1 = require("../../logic/operations/NodeDiagnosticDirPlugin");
/**
 * This class implements phased commands which are run individually for each project in the repo,
 * possibly in parallel, and which may define multiple phases.
 *
 * @remarks
 * Phased commands can be defined via common/config/command-line.json.  Rush's predefined "build"
 * and "rebuild" commands are also modeled as phased commands with a single phase that invokes the npm
 * "build" script for each project.
 */
class PhasedScriptAction extends BaseScriptAction_1.BaseScriptAction {
    constructor(options) {
        var _a;
        super(options);
        this._enableParallelism = options.enableParallelism;
        this._isIncrementalBuildAllowed = options.incremental;
        this._disableBuildCache = options.disableBuildCache;
        this._originalPhases = options.originalPhases;
        this._initialPhases = options.initialPhases;
        this._watchPhases = options.watchPhases;
        this._watchDebounceMs = (_a = options.watchDebounceMs) !== null && _a !== void 0 ? _a : RushConstants_1.RushConstants.defaultWatchDebounceMs;
        this._alwaysWatch = options.alwaysWatch;
        this._alwaysInstall = options.alwaysInstall;
        this._runsBeforeInstall = false;
        this._knownPhases = options.phases;
        this.hooks = new PhasedCommandHooks_1.PhasedCommandHooks();
        const terminal = new terminal_1.Terminal(this.rushSession.terminalProvider);
        this._terminal = terminal;
        // Generates the default operation graph
        new PhasedOperationPlugin_1.PhasedOperationPlugin().apply(this.hooks);
        // Splices in sharded phases to the operation graph.
        new ShardedPhaseOperationPlugin_1.ShardedPhasedOperationPlugin().apply(this.hooks);
        // Applies the Shell Operation Runner to selected operations
        new ShellOperationRunnerPlugin_1.ShellOperationRunnerPlugin().apply(this.hooks);
        new WeightedOperationPlugin_1.WeightedOperationPlugin().apply(this.hooks);
        new ValidateOperationsPlugin_1.ValidateOperationsPlugin(terminal).apply(this.hooks);
        if (this._enableParallelism) {
            this._parallelismParameter = this.defineStringParameter({
                parameterLongName: '--parallelism',
                parameterShortName: '-p',
                argumentName: 'COUNT',
                environmentVariable: EnvironmentConfiguration_1.EnvironmentVariableNames.RUSH_PARALLELISM,
                description: 'Specifies the maximum number of concurrent processes to launch during a build.' +
                    ' The COUNT should be a positive integer, a percentage value (eg. "50%%") or the word "max"' +
                    ' to specify a count that is equal to the number of CPU cores. If this parameter is omitted,' +
                    ' then the default value depends on the operating system and number of CPU cores.'
            });
        }
        this._timelineParameter = this.defineFlagParameter({
            parameterLongName: '--timeline',
            description: 'After the build is complete, print additional statistics and CPU usage information,' +
                ' including an ASCII chart of the start and stop times for each operation.'
        });
        this._cobuildPlanParameter = this.defineFlagParameter({
            parameterLongName: '--log-cobuild-plan',
            description: '(EXPERIMENTAL) Before the build starts, log information about the cobuild state. This will include information about ' +
                'clusters and the projects that are part of each cluster.'
        });
        this._selectionParameters = new SelectionParameterSet_1.SelectionParameterSet(this.rushConfiguration, this, {
            gitOptions: {
                // Include lockfile processing since this expands the selection, and we need to select
                // at least the same projects selected with the same query to "rush build"
                includeExternalDependencies: true,
                // Enable filtering to reduce evaluation cost
                enableFiltering: true
            },
            includeSubspaceSelector: false
        });
        this._verboseParameter = this.defineFlagParameter({
            parameterLongName: '--verbose',
            parameterShortName: '-v',
            description: 'Display the logs during the build, rather than just displaying the build status summary'
        });
        this._includePhaseDeps = this.defineFlagParameter({
            parameterLongName: '--include-phase-deps',
            description: 'If the selected projects are "unsafe" (missing some dependencies), add the minimal set of phase dependencies. For example, ' +
                `"--from A" normally might include the "_phase:test" phase for A's dependencies, even though changes to A can't break those tests. ` +
                `Using "--impacted-by A --include-phase-deps" avoids that work by performing "_phase:test" only for downstream projects.`
        });
        if (this._isIncrementalBuildAllowed) {
            this._changedProjectsOnly = this.defineFlagParameter({
                parameterLongName: '--changed-projects-only',
                parameterShortName: '-c',
                description: 'Normally the incremental build logic will rebuild changed projects as well as' +
                    ' any projects that directly or indirectly depend on a changed project. Specify "--changed-projects-only"' +
                    ' to ignore dependent projects, only rebuilding those projects whose files were changed.' +
                    ' Note that this parameter is "unsafe"; it is up to the developer to ensure that the ignored projects' +
                    ' are okay to ignore.'
            });
        }
        this._ignoreHooksParameter = this.defineFlagParameter({
            parameterLongName: '--ignore-hooks',
            description: `Skips execution of the "eventHooks" scripts defined in ${RushConstants_1.RushConstants.rushJsonFilename}. ` +
                'Make sure you know what you are skipping.'
        });
        if (this._watchPhases.size > 0 && !this._alwaysWatch) {
            // Only define the parameter if it has an effect.
            this._watchParameter = this.defineFlagParameter({
                parameterLongName: '--watch',
                description: `Starts a file watcher after initial execution finishes. Will run the following phases on affected projects: ${Array.from(this._watchPhases, (phase) => phase.name).join(', ')}`
            });
        }
        // If `this._alwaysInstall === undefined`, Rush does not define the parameter
        // but a repository may still define a custom parameter with the same name.
        if (this._alwaysInstall === false) {
            this._installParameter = this.defineFlagParameter({
                parameterLongName: '--install',
                description: 'Normally a phased command expects "rush install" to have been manually run first. If this flag is specified, ' +
                    'Rush will automatically perform an install before processing the current command.'
            });
        }
        if (this._alwaysInstall !== undefined) {
            this._variantParameter = this.defineStringParameter(Variants_1.VARIANT_PARAMETER);
        }
        if (this._watchPhases.size > 0 &&
            this.rushConfiguration.experimentsConfiguration.configuration.useIPCScriptsInWatchMode) {
            this._noIPCParameter = this.defineFlagParameter({
                parameterLongName: '--no-ipc',
                description: 'Disables the IPC feature for the current command (if applicable to selected operations). Operations will not look for a ":ipc" suffixed script.' +
                    'This feature only applies in watch mode and is enabled by default.'
            });
        }
        this._nodeDiagnosticDirParameter = this.defineStringParameter({
            parameterLongName: '--node-diagnostic-dir',
            argumentName: 'DIRECTORY',
            description: 'Specifies the directory where Node.js diagnostic reports will be written. ' +
                'This directory will contain a subdirectory for each project and phase.'
        });
        this.defineScriptParameters();
        for (const [{ associatedPhases }, tsCommandLineParameter] of this.customParameters) {
            if (associatedPhases) {
                for (const phaseName of associatedPhases) {
                    const phase = this._knownPhases.get(phaseName);
                    if (!phase) {
                        throw new node_core_library_1.InternalError(`Could not find a phase matching ${phaseName}.`);
                    }
                    phase.associatedParameters.add(tsCommandLineParameter);
                }
            }
        }
    }
    async runAsync() {
        var _a, _b, _c, _d, _e, _f, _g, _h, _j, _k;
        if (this._alwaysInstall || ((_a = this._installParameter) === null || _a === void 0 ? void 0 : _a.value)) {
            const { doBasicInstallAsync } = await Promise.resolve().then(() => __importStar(require(
            /* webpackChunkName: 'doBasicInstallAsync' */
            '../../logic/installManager/doBasicInstallAsync')));
            const variant = await (0, Variants_1.getVariantAsync)(this._variantParameter, this.rushConfiguration, true);
            await doBasicInstallAsync({
                terminal: this._terminal,
                rushConfiguration: this.rushConfiguration,
                rushGlobalFolder: this.rushGlobalFolder,
                isDebug: this.parser.isDebug,
                variant,
                beforeInstallAsync: (subspace) => this.rushSession.hooks.beforeInstall.promise(this, subspace, variant),
                afterInstallAsync: (subspace) => this.rushSession.hooks.afterInstall.promise(this, subspace, variant),
                // Eventually we may want to allow a subspace to be selected here
                subspace: this.rushConfiguration.defaultSubspace
            });
        }
        if (!this._runsBeforeInstall) {
            // TODO: Replace with last-install.flag when "rush link" and "rush unlink" are removed
            const lastLinkFlag = new FlagFile_1.FlagFile(this.rushConfiguration.defaultSubspace.getSubspaceTempFolderPath(), RushConstants_1.RushConstants.lastLinkFlagFilename, {});
            // Only check for a valid link flag when subspaces is not enabled
            if (!(await lastLinkFlag.isValidAsync()) && !this.rushConfiguration.subspacesFeatureEnabled) {
                const useWorkspaces = this.rushConfiguration.pnpmOptions && this.rushConfiguration.pnpmOptions.useWorkspaces;
                if (useWorkspaces) {
                    throw new Error('Link flag invalid.\nDid you run "rush install" or "rush update"?');
                }
                else {
                    throw new Error('Link flag invalid.\nDid you run "rush link"?');
                }
            }
        }
        this._doBeforeTask();
        // if this is parallelizable, then use the value from the flag (undefined or a number),
        // if parallelism is not enabled, then restrict to 1 core
        const parallelism = this._enableParallelism
            ? (0, ParseParallelism_1.parseParallelism)((_b = this._parallelismParameter) === null || _b === void 0 ? void 0 : _b.value)
            : 1;
        const terminal = this._terminal;
        const stopwatch = Stopwatch_1.Stopwatch.start();
        const showTimeline = this._timelineParameter ? this._timelineParameter.value : false;
        if (showTimeline) {
            const { ConsoleTimelinePlugin } = await Promise.resolve().then(() => __importStar(require(
            /* webpackChunkName: 'ConsoleTimelinePlugin' */
            '../../logic/operations/ConsoleTimelinePlugin')));
            new ConsoleTimelinePlugin(terminal).apply(this.hooks);
        }
        const includePhaseDeps = (_d = (_c = this._includePhaseDeps) === null || _c === void 0 ? void 0 : _c.value) !== null && _d !== void 0 ? _d : false;
        const diagnosticDir = this._nodeDiagnosticDirParameter.value;
        if (diagnosticDir) {
            new NodeDiagnosticDirPlugin_1.NodeDiagnosticDirPlugin({
                diagnosticDir
            }).apply(this.hooks);
        }
        // Enable the standard summary
        new OperationResultSummarizerPlugin_1.OperationResultSummarizerPlugin(terminal).apply(this.hooks);
        const { hooks: sessionHooks } = this.rushSession;
        if (sessionHooks.runAnyPhasedCommand.isUsed()) {
            // Avoid the cost of compiling the hook if it wasn't tapped.
            await sessionHooks.runAnyPhasedCommand.promise(this);
        }
        const hookForAction = sessionHooks.runPhasedCommand.get(this.actionName);
        if (hookForAction) {
            // Run the more specific hook for a command with this name after the general hook
            await hookForAction.promise(this);
        }
        const isQuietMode = !this._verboseParameter.value;
        const changedProjectsOnly = !!((_e = this._changedProjectsOnly) === null || _e === void 0 ? void 0 : _e.value);
        let buildCacheConfiguration;
        let cobuildConfiguration;
        if (!this._disableBuildCache) {
            buildCacheConfiguration = await BuildCacheConfiguration_1.BuildCacheConfiguration.tryLoadAsync(terminal, this.rushConfiguration, this.rushSession);
            cobuildConfiguration = await CobuildConfiguration_1.CobuildConfiguration.tryLoadAsync(terminal, this.rushConfiguration, this.rushSession);
            await (cobuildConfiguration === null || cobuildConfiguration === void 0 ? void 0 : cobuildConfiguration.createLockProviderAsync(terminal));
        }
        try {
            const projectSelection = await this._selectionParameters.getSelectedProjectsAsync(terminal);
            if (!projectSelection.size) {
                terminal.writeLine(terminal_1.Colorize.yellow(`The command line selection parameters did not match any projects.`));
                return;
            }
            const isWatch = ((_f = this._watchParameter) === null || _f === void 0 ? void 0 : _f.value) || this._alwaysWatch;
            if (isWatch && ((_g = this._noIPCParameter) === null || _g === void 0 ? void 0 : _g.value) === false) {
                new (await Promise.resolve().then(() => __importStar(require(
                /* webpackChunkName: 'IPCOperationRunnerPlugin' */ '../../logic/operations/IPCOperationRunnerPlugin')))).IPCOperationRunnerPlugin().apply(this.hooks);
            }
            const customParametersByName = new Map();
            for (const [configParameter, parserParameter] of this.customParameters) {
                customParametersByName.set(configParameter.longName, parserParameter);
            }
            if (buildCacheConfiguration === null || buildCacheConfiguration === void 0 ? void 0 : buildCacheConfiguration.buildCacheEnabled) {
                terminal.writeVerboseLine(`Incremental strategy: cache restoration`);
                new CacheableOperationPlugin_1.CacheableOperationPlugin({
                    allowWarningsInSuccessfulBuild: !!this.rushConfiguration.experimentsConfiguration.configuration
                        .buildCacheWithAllowWarningsInSuccessfulBuild,
                    buildCacheConfiguration,
                    cobuildConfiguration,
                    terminal
                }).apply(this.hooks);
            }
            else if (!this._disableBuildCache) {
                terminal.writeVerboseLine(`Incremental strategy: output preservation`);
                // Explicitly disabling the build cache also disables legacy skip detection.
                new LegacySkipPlugin_1.LegacySkipPlugin({
                    allowWarningsInSuccessfulBuild: this.rushConfiguration.experimentsConfiguration.configuration
                        .buildSkipWithAllowWarningsInSuccessfulBuild,
                    terminal,
                    changedProjectsOnly,
                    isIncrementalBuildAllowed: this._isIncrementalBuildAllowed
                }).apply(this.hooks);
            }
            else {
                terminal.writeVerboseLine(`Incremental strategy: none (full rebuild)`);
            }
            const showBuildPlan = (_j = (_h = this._cobuildPlanParameter) === null || _h === void 0 ? void 0 : _h.value) !== null && _j !== void 0 ? _j : false;
            if (showBuildPlan) {
                if (!(buildCacheConfiguration === null || buildCacheConfiguration === void 0 ? void 0 : buildCacheConfiguration.buildCacheEnabled)) {
                    throw new Error('You must have build cache enabled to use this option.');
                }
                const { BuildPlanPlugin } = await Promise.resolve().then(() => __importStar(require('../../logic/operations/BuildPlanPlugin')));
                new BuildPlanPlugin(terminal).apply(this.hooks);
            }
            const { configuration: experiments } = this.rushConfiguration.experimentsConfiguration;
            if (((_k = this.rushConfiguration) === null || _k === void 0 ? void 0 : _k.isPnpm) && (experiments === null || experiments === void 0 ? void 0 : experiments.usePnpmSyncForInjectedDependencies)) {
                const { PnpmSyncCopyOperationPlugin } = await Promise.resolve().then(() => __importStar(require('../../logic/operations/PnpmSyncCopyOperationPlugin')));
                new PnpmSyncCopyOperationPlugin(terminal).apply(this.hooks);
            }
            const relevantProjects = Selection_1.Selection.expandAllDependencies(projectSelection);
            const projectConfigurations = this
                ._runsBeforeInstall
                ? new Map()
                : await RushProjectConfiguration_1.RushProjectConfiguration.tryLoadForProjectsAsync(relevantProjects, terminal);
            const initialCreateOperationsContext = {
                buildCacheConfiguration,
                cobuildConfiguration,
                customParameters: customParametersByName,
                isIncrementalBuildAllowed: this._isIncrementalBuildAllowed,
                isInitial: true,
                isWatch,
                rushConfiguration: this.rushConfiguration,
                phaseOriginal: new Set(this._originalPhases),
                phaseSelection: new Set(this._initialPhases),
                includePhaseDeps,
                projectSelection,
                projectConfigurations,
                projectsInUnknownState: projectSelection
            };
            const executionManagerOptions = {
                quietMode: isQuietMode,
                debugMode: this.parser.isDebug,
                parallelism,
                changedProjectsOnly,
                beforeExecuteOperationAsync: async (record) => {
                    return await this.hooks.beforeExecuteOperation.promise(record);
                },
                afterExecuteOperationAsync: async (record) => {
                    await this.hooks.afterExecuteOperation.promise(record);
                },
                createEnvironmentForOperation: this.hooks.createEnvironmentForOperation.isUsed()
                    ? (record) => {
                        return this.hooks.createEnvironmentForOperation.call(Object.assign({}, process.env), record);
                    }
                    : undefined,
                onOperationStatusChangedAsync: (record) => {
                    this.hooks.onOperationStatusChanged.call(record);
                }
            };
            const initialInternalOptions = {
                initialCreateOperationsContext,
                executionManagerOptions,
                stopwatch,
                terminal
            };
            const internalOptions = await this._runInitialPhasesAsync(initialInternalOptions);
            if (isWatch) {
                if (buildCacheConfiguration) {
                    // Cache writes are not supported during watch mode, only reads.
                    buildCacheConfiguration.cacheWriteEnabled = false;
                }
                await this._runWatchPhasesAsync(internalOptions);
            }
        }
        finally {
            await (cobuildConfiguration === null || cobuildConfiguration === void 0 ? void 0 : cobuildConfiguration.destroyLockProviderAsync());
        }
    }
    async _runInitialPhasesAsync(options) {
        const { initialCreateOperationsContext, executionManagerOptions: partialExecutionManagerOptions, stopwatch, terminal } = options;
        const { projectConfigurations } = initialCreateOperationsContext;
        const { projectSelection } = initialCreateOperationsContext;
        const operations = await this.hooks.createOperations.promise(new Set(), initialCreateOperationsContext);
        terminal.write('Analyzing repo state... ');
        const repoStateStopwatch = new Stopwatch_1.Stopwatch();
        repoStateStopwatch.start();
        const analyzer = new ProjectChangeAnalyzer_1.ProjectChangeAnalyzer(this.rushConfiguration);
        const getInputsSnapshotAsync = await analyzer._tryGetSnapshotProviderAsync(projectConfigurations, terminal, 
        // We need to include all dependencies, otherwise build cache id calculation will be incorrect
        Selection_1.Selection.expandAllDependencies(projectSelection));
        const initialSnapshot = await (getInputsSnapshotAsync === null || getInputsSnapshotAsync === void 0 ? void 0 : getInputsSnapshotAsync());
        repoStateStopwatch.stop();
        terminal.writeLine(`DONE (${repoStateStopwatch.toString()})`);
        terminal.writeLine();
        const initialExecuteOperationsContext = Object.assign(Object.assign({}, initialCreateOperationsContext), { inputsSnapshot: initialSnapshot });
        const executionManagerOptions = Object.assign(Object.assign({}, partialExecutionManagerOptions), { beforeExecuteOperationsAsync: async (records) => {
                await this.hooks.beforeExecuteOperations.promise(records, initialExecuteOperationsContext);
            } });
        const initialOptions = {
            executeOperationsContext: initialExecuteOperationsContext,
            ignoreHooks: false,
            operations,
            stopwatch,
            executionManagerOptions,
            terminal
        };
        await this._executeOperationsAsync(initialOptions);
        return Object.assign(Object.assign({}, options), { executionManagerOptions,
            getInputsSnapshotAsync,
            initialSnapshot });
    }
    _registerWatchModeInterface(projectWatcher) {
        const toggleWatcherKey = 'w';
        const buildOnceKey = 'b';
        const invalidateKey = 'i';
        const shutdownKey = 'x';
        const terminal = this._terminal;
        projectWatcher.setPromptGenerator((isPaused) => {
            var _a;
            const promptLines = [
                `  Press <${toggleWatcherKey}> to ${isPaused ? 'resume' : 'pause'}.`,
                `  Press <${invalidateKey}> to invalidate all projects.`
            ];
            if (isPaused) {
                promptLines.push(`  Press <${buildOnceKey}> to build once.`);
            }
            if (((_a = this._noIPCParameter) === null || _a === void 0 ? void 0 : _a.value) === false) {
                promptLines.push(`  Press <${shutdownKey}> to reset child processes.`);
            }
            return promptLines;
        });
        process.stdin.setRawMode(true);
        process.stdin.resume();
        process.stdin.setEncoding('utf8');
        process.stdin.on('data', (key) => {
            switch (key) {
                case toggleWatcherKey:
                    if (projectWatcher.isPaused) {
                        projectWatcher.resume();
                    }
                    else {
                        projectWatcher.pause();
                    }
                    break;
                case buildOnceKey:
                    if (projectWatcher.isPaused) {
                        projectWatcher.clearStatus();
                        terminal.writeLine(`Building once...`);
                        projectWatcher.resume();
                        projectWatcher.pause();
                    }
                    break;
                case invalidateKey:
                    projectWatcher.clearStatus();
                    terminal.writeLine(`Invalidating all operations...`);
                    projectWatcher.invalidateAll('manual trigger');
                    if (!projectWatcher.isPaused) {
                        projectWatcher.resume();
                    }
                    break;
                case shutdownKey:
                    projectWatcher.clearStatus();
                    terminal.writeLine(`Shutting down long-lived child processes...`);
                    // TODO: Inject this promise into the execution queue somewhere so that it gets waited on between runs
                    void this.hooks.shutdownAsync.promise();
                    break;
                case '\u0003':
                    process.kill(process.pid, 'SIGINT');
                    break;
            }
        });
    }
    /**
     * Runs the command in watch mode. Fundamentally is a simple loop:
     * 1) Wait for a change to one or more projects in the selection
     * 2) Invoke the command on the changed projects, and, if applicable, impacted projects
     *    Uses the same algorithm as --impacted-by
     * 3) Goto (1)
     */
    async _runWatchPhasesAsync(options) {
        const { getInputsSnapshotAsync, initialSnapshot, initialCreateOperationsContext, executionManagerOptions, stopwatch, terminal } = options;
        const phaseOriginal = new Set(this._watchPhases);
        const phaseSelection = new Set(this._watchPhases);
        const { projectSelection: projectsToWatch } = initialCreateOperationsContext;
        if (!getInputsSnapshotAsync || !initialSnapshot) {
            terminal.writeErrorLine(`Cannot watch for changes if the Rush repo is not in a Git repository, exiting.`);
            throw new node_core_library_1.AlreadyReportedError();
        }
        // Use async import so that we don't pay the cost for sync builds
        const { ProjectWatcher } = await Promise.resolve().then(() => __importStar(require(
        /* webpackChunkName: 'ProjectWatcher' */
        '../../logic/ProjectWatcher')));
        const projectWatcher = new ProjectWatcher({
            getInputsSnapshotAsync,
            initialSnapshot,
            debounceMs: this._watchDebounceMs,
            rushConfiguration: this.rushConfiguration,
            projectsToWatch,
            terminal
        });
        // Ensure process.stdin allows interactivity before using TTY-only APIs
        if (process.stdin.isTTY) {
            this._registerWatchModeInterface(projectWatcher);
        }
        const onWaitingForChanges = () => {
            // Allow plugins to display their own messages when waiting for changes.
            this.hooks.waitingForChanges.call();
            // Report so that the developer can always see that it is in watch mode as the latest console line.
            terminal.writeLine(`Watching for changes to ${projectsToWatch.size} ${projectsToWatch.size === 1 ? 'project' : 'projects'}. Press Ctrl+C to exit.`);
        };
        function invalidateOperation(operation, reason) {
            const { associatedProject } = operation;
            if (associatedProject) {
                // Since ProjectWatcher only tracks entire projects, widen the operation to its project
                // Revisit when migrating to @rushstack/operation-graph and we have a long-lived operation graph
                projectWatcher.invalidateProject(associatedProject, `${operation.name} (${reason})`);
            }
        }
        // Loop until Ctrl+C
        // eslint-disable-next-line no-constant-condition
        while (true) {
            // On the initial invocation, this promise will return immediately with the full set of projects
            const { changedProjects, inputsSnapshot: state } = await projectWatcher.waitForChangeAsync(onWaitingForChanges);
            if (stopwatch.state === Stopwatch_1.StopwatchState.Stopped) {
                // Clear and reset the stopwatch so that we only report time from a single execution at a time
                stopwatch.reset();
                stopwatch.start();
            }
            terminal.writeLine(`Detected changes in ${changedProjects.size} project${changedProjects.size === 1 ? '' : 's'}:`);
            const names = [...changedProjects].map((x) => x.packageName).sort();
            for (const name of names) {
                terminal.writeLine(`    ${terminal_1.Colorize.cyan(name)}`);
            }
            // Account for consumer relationships
            const executeOperationsContext = Object.assign(Object.assign({}, initialCreateOperationsContext), { isInitial: false, inputsSnapshot: state, projectsInUnknownState: changedProjects, phaseOriginal,
                phaseSelection,
                invalidateOperation });
            const operations = await this.hooks.createOperations.promise(new Set(), executeOperationsContext);
            const executeOptions = {
                executeOperationsContext,
                // For now, don't run pre-build or post-build in watch mode
                ignoreHooks: true,
                operations,
                stopwatch,
                executionManagerOptions: Object.assign(Object.assign({}, executionManagerOptions), { beforeExecuteOperationsAsync: async (records) => {
                        await this.hooks.beforeExecuteOperations.promise(records, executeOperationsContext);
                    } }),
                terminal
            };
            try {
                // Delegate the the underlying command, for only the projects that need reprocessing
                await this._executeOperationsAsync(executeOptions);
            }
            catch (err) {
                // In watch mode, we want to rebuild even if the original build failed.
                if (!(err instanceof node_core_library_1.AlreadyReportedError)) {
                    throw err;
                }
            }
        }
    }
    /**
     * Runs a set of operations and reports the results.
     */
    async _executeOperationsAsync(options) {
        const { executionManagerOptions, ignoreHooks, operations, stopwatch, terminal } = options;
        const executionManager = new OperationExecutionManager_1.OperationExecutionManager(operations, executionManagerOptions);
        const { isInitial, isWatch, cobuildConfiguration } = options.executeOperationsContext;
        let success = false;
        let result;
        try {
            result = await executionManager.executeAsync();
            success = result.status === OperationStatus_1.OperationStatus.Success;
            await this.hooks.afterExecuteOperations.promise(result, options.executeOperationsContext);
            stopwatch.stop();
            const message = `rush ${this.actionName} (${stopwatch.toString()})`;
            if (result.status === OperationStatus_1.OperationStatus.Success) {
                terminal.writeLine(terminal_1.Colorize.green(message));
            }
            else {
                terminal.writeLine(message);
            }
        }
        catch (error) {
            success = false;
            stopwatch.stop();
            if (error instanceof node_core_library_1.AlreadyReportedError) {
                terminal.writeLine(`rush ${this.actionName} (${stopwatch.toString()})`);
            }
            else {
                if (error && error.message) {
                    if (this.parser.isDebug) {
                        terminal.writeErrorLine('Error: ' + error.stack);
                    }
                    else {
                        terminal.writeErrorLine('Error: ' + error.message);
                    }
                }
                terminal.writeErrorLine(terminal_1.Colorize.red(`rush ${this.actionName} - Errors! (${stopwatch.toString()})`));
            }
        }
        if (!ignoreHooks) {
            this._doAfterTask();
        }
        if (this.parser.telemetry) {
            const jsonOperationResults = {};
            const extraData = Object.assign(Object.assign(Object.assign({}, this._selectionParameters.getTelemetry()), this.getParameterStringMap()), { isWatch,
                // Fields specific to the current operation set
                isInitial, countAll: 0, countSuccess: 0, countSuccessWithWarnings: 0, countFailure: 0, countBlocked: 0, countFromCache: 0, countSkipped: 0, countNoOp: 0 });
            if (result) {
                const { operationResults } = result;
                const nonSilentDependenciesByOperation = new Map();
                function getNonSilentDependencies(operation) {
                    let realDependencies = nonSilentDependenciesByOperation.get(operation);
                    if (!realDependencies) {
                        realDependencies = new Set();
                        nonSilentDependenciesByOperation.set(operation, realDependencies);
                        for (const dependency of operation.dependencies) {
                            const dependencyRecord = operationResults.get(dependency);
                            if (dependencyRecord === null || dependencyRecord === void 0 ? void 0 : dependencyRecord.silent) {
                                for (const deepDependency of getNonSilentDependencies(dependency)) {
                                    realDependencies.add(deepDependency);
                                }
                            }
                            else {
                                realDependencies.add(dependency.name);
                            }
                        }
                    }
                    return realDependencies;
                }
                for (const [operation, operationResult] of operationResults) {
                    if (operationResult.silent) {
                        // Architectural operation. Ignore.
                        continue;
                    }
                    const { startTime, endTime } = operationResult.stopwatch;
                    jsonOperationResults[operation.name] = {
                        startTimestampMs: startTime,
                        endTimestampMs: endTime,
                        nonCachedDurationMs: operationResult.nonCachedDurationMs,
                        wasExecutedOnThisMachine: !operationResult.cobuildRunnerId ||
                            operationResult.cobuildRunnerId === (cobuildConfiguration === null || cobuildConfiguration === void 0 ? void 0 : cobuildConfiguration.cobuildRunnerId),
                        result: operationResult.status,
                        dependencies: Array.from(getNonSilentDependencies(operation)).sort()
                    };
                    extraData.countAll++;
                    switch (operationResult.status) {
                        case OperationStatus_1.OperationStatus.Success:
                            extraData.countSuccess++;
                            break;
                        case OperationStatus_1.OperationStatus.SuccessWithWarning:
                            extraData.countSuccessWithWarnings++;
                            break;
                        case OperationStatus_1.OperationStatus.Failure:
                            extraData.countFailure++;
                            break;
                        case OperationStatus_1.OperationStatus.Blocked:
                            extraData.countBlocked++;
                            break;
                        case OperationStatus_1.OperationStatus.FromCache:
                            extraData.countFromCache++;
                            break;
                        case OperationStatus_1.OperationStatus.Skipped:
                            extraData.countSkipped++;
                            break;
                        case OperationStatus_1.OperationStatus.NoOp:
                            extraData.countNoOp++;
                            break;
                        default:
                            // Do nothing.
                            break;
                    }
                }
            }
            const logEntry = {
                name: this.actionName,
                durationInSeconds: stopwatch.duration,
                result: success ? 'Succeeded' : 'Failed',
                extraData,
                operationResults: jsonOperationResults
            };
            this.hooks.beforeLog.call(logEntry);
            this.parser.telemetry.log(logEntry);
            this.parser.flushTelemetry();
        }
        if (!success && !isWatch) {
            throw new node_core_library_1.AlreadyReportedError();
        }
    }
    _doBeforeTask() {
        if (this.actionName !== RushConstants_1.RushConstants.buildCommandName &&
            this.actionName !== RushConstants_1.RushConstants.rebuildCommandName) {
            // Only collects information for built-in commands like build or rebuild.
            return;
        }
        SetupChecks_1.SetupChecks.validate(this.rushConfiguration);
        this.eventHooksManager.handle(EventHooks_1.Event.preRushBuild, this.parser.isDebug, this._ignoreHooksParameter.value);
    }
    _doAfterTask() {
        if (this.actionName !== RushConstants_1.RushConstants.buildCommandName &&
            this.actionName !== RushConstants_1.RushConstants.rebuildCommandName) {
            // Only collects information for built-in commands like build or rebuild.
            return;
        }
        this.eventHooksManager.handle(EventHooks_1.Event.postRushBuild, this.parser.isDebug, this._ignoreHooksParameter.value);
    }
}
exports.PhasedScriptAction = PhasedScriptAction;
