// @microsoft/spfx-heft-plugins/lib-esm/utilities/UrlUtilities.js
import * as url from 'url';
/**
 * Utility functions for working with urls
 * @public
 */ export class UrlUtilities {
    static joinUrlSegments(...segments) {
        if (!segments || segments.length === 0) {
            return '';
        }
        segments = segments.filter((value)=>!!value);
        let result = segments[0];
        for(let i = 1; i < segments.length; i++){
            result = url.resolve(`${UrlUtilities.trimTrailingSlashes(result)}/`, UrlUtilities.trimLeadingSlashes(segments[i]));
        }
        return UrlUtilities.trimTrailingSlashes(result);
    }
    static ensureTrailingSlash(str) {
        if (!str) {
            return str;
        }
        return str.replace(/\/*$/, '/');
    }
    static isUrl(path) {
        /* Matches the following:
     * http://contoso.org/foo.js
     * https://contoso
     * http:/\foo.bar
     * https:\/foo.bar
     * http:\\foo.bar
     * https:\\foo.bar
     * HTTp:\\foo.bar
     */ const urlRegex = /([hH][tT]{2}[pP]s?):[/\\]{2}/;
        return !!path.match(urlRegex);
    }
    static trimTrailingSlashes(str) {
        if (!str) {
            return '';
        }
        return str.replace(/\/+$/, '');
    }
    static trimLeadingSlashes(str) {
        if (!str) {
            return '';
        }
        return str.replace(/^\/+/, '');
    }
}

//#sourceMappingUrl=./UrlUtilities.js.map
