// @fluentui/utilities/lib-amd/safeRequestAnimationFrame.test.js
define(["require", "exports", "tslib", "react", "./safeRequestAnimationFrame", "react-test-renderer"], function (require, exports, tslib_1, React, safeRequestAnimationFrame_1, renderer) {
    "use strict";
    Object.defineProperty(exports, "__esModule", { value: true });
    describe('safeRequestAnimationFrame', function () {
        var rafCalled = false;
        var Foo = /** @class */ (function (_super) {
            tslib_1.__extends(Foo, _super);
            function Foo(props) {
                var _this = _super.call(this, props) || this;
                _this._raf = (0, safeRequestAnimationFrame_1.safeRequestAnimationFrame)(_this);
                return _this;
            }
            Foo.prototype.render = function () {
                return React.createElement("div", null, "Hello");
            };
            Foo.prototype.componentDidMount = function () {
                this._raf(function () { return (rafCalled = true); });
            };
            return Foo;
        }(React.Component));
        beforeEach(function () {
            rafCalled = false;
            jest.useFakeTimers();
        });
        afterEach(function () {
            jest.runOnlyPendingTimers();
            jest.useRealTimers();
        });
        it('can request animation frame', function () {
            renderer.create(React.createElement(Foo, null));
            expect(rafCalled).toEqual(false);
            jest.runOnlyPendingTimers();
            expect(rafCalled).toEqual(true);
        });
        it('can cancel request animation frame', function () {
            var component = renderer.create(React.createElement(Foo, null));
            expect(rafCalled).toEqual(false);
            component.unmount();
            jest.runOnlyPendingTimers();
            expect(rafCalled).toEqual(false);
        });
    });
});
