// @microsoft/spfx-heft-plugins/lib-commonjs/plugins/packageSolutionPlugin/PackageSolutionPlugin.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return PackageSolutionPlugin;
    }
});
const _PluginNames = require("../PluginNames");
const _SolutionPackager = require("./SolutionPackager");
class PackageSolutionPlugin {
    apply(taskSession, heftConfiguration) {
        const logger = taskSession.logger;
        taskSession.hooks.run.tapPromise(_PluginNames.PackageSolutionPlugin, async ()=>{
            const prettyPrintXmlParameter = taskSession.parameters.getFlagParameter('--pretty-print-xml');
            const noSpppkgParameter = taskSession.parameters.getFlagParameter('--no-spppkg');
            const solutionPackager = new _SolutionPackager.SolutionPackager({
                buildFolder: heftConfiguration.buildFolderPath,
                terminal: heftConfiguration.globalTerminal,
                emitError: logger.emitError.bind(logger),
                emitWarning: logger.emitWarning.bind(logger)
            });
            await solutionPackager.packageSolutionAsync(taskSession.parameters.production, prettyPrintXmlParameter.value, noSpppkgParameter.value);
        });
    }
}

//#sourceMappingUrl=./PackageSolutionPlugin.js.map
