// @pnp/spfx-property-controls/lib/propertyFields/collectionData/PropertyFieldCollectionDataHost.module.scss.js
require("./PropertyFieldCollectionDataHost.module.css");
const styles = {
    collectionData: 'collectionData_f8375039',
    noCollectionData: 'noCollectionData_f8375039',
    panelActions: 'panelActions_f8375039',
    required: 'required_f8375039',
    addBtn: 'addBtn_f8375039',
    addBtnDisabled: 'addBtnDisabled_f8375039',
    inputField: 'inputField_f8375039',
    numberField: 'numberField_f8375039',
    invalidField: 'invalidField_f8375039',
    iconField: 'iconField_f8375039',
    iconPicker: 'iconPicker_f8375039',
    requiredField: 'requiredField_f8375039',
    colorField: 'colorField_f8375039',
    colorIndicator: 'colorIndicator_f8375039',
    collectionDataField: 'collectionDataField_f8375039',
    table: 'table_f8375039',
    tableRow: 'tableRow_f8375039',
    tableFooter: 'tableFooter_f8375039',
    tableCell: 'tableCell_f8375039',
    errorCallout: 'errorCallout_f8375039',
    errorCalloutLink: 'errorCalloutLink_f8375039',
    errorMsgs: 'errorMsgs_f8375039',
    tableHead: 'tableHead_f8375039'
};
export default styles;
