// @microsoft/sp-loader/lib-commonjs/tiny/common/killSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isGuestUserForAnonymousAndEmailAuthenticatedKSActive = exports.isUseAADTenantUrlFromContextKSActivated = exports.isDoNotSetCallbackForSharePointMobileKSActive = exports.isVivaHomeReloadWithBypassKSActive = exports.isVivaHomeReloadKSActive = exports.isDoNotSetCallbackForVivaMobileKSActive = exports.isCatchDebugEvalErrorsKSActivated = exports.isBotACEManifestLoadLogicRemovedKSActivated = exports.isAlwaysInitOboConfigKSActive = exports.isDoNotApplyCustomSpfxPrincipalToIsolationKSActive = exports.isTeamsBrowserHostedKSActive = exports.isGetSessionIdFromQueryParametersKSActivated = exports.overrideWorkloadKSActivated = exports.useOBOInWebForSecureBrokerKSActivated = exports.overrideXMLHttpRequestKSActivated = exports.notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain = exports.traceRequireLoaderErrorKSActivated = exports.isVersionMismatchThrowKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
function isVersionMismatchThrowKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('676ca369-31fe-4888-bad3-5ecfc6d828cd' /* '3/9/22', 'Throw when requesting a higher version than a pinned manifest' */);
}
exports.isVersionMismatchThrowKSActivated = isVersionMismatchThrowKSActivated;
function traceRequireLoaderErrorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b36bb138-515a-420a-9d65-e09d90cc855f' /* '6/15/22', 'trace loader errors to the console for easier debugging.' */);
}
exports.traceRequireLoaderErrorKSActivated = traceRequireLoaderErrorKSActivated;
function notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain() {
    return sp_core_library_1._SPKillSwitch.isActivated('5910ea14-bd58-4817-90dd-933e0274fe1d' /* '5/8/22', 'Use securebroker app id on securebroker domain' */);
}
exports.notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain = notUseFirstPartyIsolatedAppIdOnSecurebrokerDomain;
function overrideXMLHttpRequestKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d51a8be3-8ec5-4e35-ada6-cceca79c52f8' /* '8/25/22', 'Override XMLHttpRequest to disable adding traceparent property.' */);
}
exports.overrideXMLHttpRequestKSActivated = overrideXMLHttpRequestKSActivated;
function useOBOInWebForSecureBrokerKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('043feae9-5a56-40ff-b175-6f7f813507ff' /* '10/26/22', 'use OBO in web for secure broker' */);
}
exports.useOBOInWebForSecureBrokerKSActivated = useOBOInWebForSecureBrokerKSActivated;
function overrideWorkloadKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('172a5a30-5ef3-4335-b2b7-d58fb63a2d37' /* '11/11/22', Override telemetry workload' */);
}
exports.overrideWorkloadKSActivated = overrideWorkloadKSActivated;
function isGetSessionIdFromQueryParametersKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ef42347e-c163-4c8b-9a66-b3cb90ae4ec5' /* '02/14/23', 'Get sessionId for AadTokenProvider configuration from query parameters' */);
}
exports.isGetSessionIdFromQueryParametersKSActivated = isGetSessionIdFromQueryParametersKSActivated;
function isTeamsBrowserHostedKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('9b9957d5-67d6-4f73-8882-11ee6bbaa066'
    /* '02/20/2024', 'Set teams hosted container' */
    );
}
exports.isTeamsBrowserHostedKSActive = isTeamsBrowserHostedKSActive;
function isDoNotApplyCustomSpfxPrincipalToIsolationKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('2dcbf2bc-78a8-4f80-9b5d-5c515063fd76'
    /* '07/20/2023', 'Disables applying custom SPFx principal to isolation' */
    );
}
exports.isDoNotApplyCustomSpfxPrincipalToIsolationKSActive = isDoNotApplyCustomSpfxPrincipalToIsolationKSActive;
// cross project KS
function isAlwaysInitOboConfigKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('7fb9acc4-19f2-446f-b592-f15c49bfd771'
    /* '10/17/2023', 'Always initialize OBO configuration' */
    );
}
exports.isAlwaysInitOboConfigKSActive = isAlwaysInitOboConfigKSActive;
function isBotACEManifestLoadLogicRemovedKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('f770be99-a2bc-4502-9e10-5afe9cc75865'
    /*'12/11/2023', 'Remove bot ACE manifest load logic' */
    );
}
exports.isBotACEManifestLoadLogicRemovedKSActivated = isBotACEManifestLoadLogicRemovedKSActivated;
function isCatchDebugEvalErrorsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('15f93dab-db8a-4b6e-882a-38e17dba9cb1'
    /* '03/15/2024', 'Catch eval exceptions during debug' */
    );
}
exports.isCatchDebugEvalErrorsKSActivated = isCatchDebugEvalErrorsKSActivated;
function isDoNotSetCallbackForVivaMobileKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('fcbb01dc-0a98-4292-9726-c339aed52550'
    /* '03/20/2024', 'Do not set callback for viva mobile' */
    );
}
exports.isDoNotSetCallbackForVivaMobileKSActive = isDoNotSetCallbackForVivaMobileKSActive;
function isVivaHomeReloadKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('1ba25635-6700-4cdf-b644-c8dfe48013d5'
    /* '6/17/24', 'temporary hack to unblock vivahome in manifest version mismatch scenario' */
    );
}
exports.isVivaHomeReloadKSActive = isVivaHomeReloadKSActive;
function isVivaHomeReloadWithBypassKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('db8d100a-e236-46bb-8dbd-b90071a46c79'
    /* '6/17/24', 'disable the full page reload from manifest version mismatch' */
    );
}
exports.isVivaHomeReloadWithBypassKSActive = isVivaHomeReloadWithBypassKSActive;
function isDoNotSetCallbackForSharePointMobileKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('65f17677-4d1e-4609-9642-712c43e395f2'
    /* '03/20/2024', 'Do not set callback for SharePoint mobile' */
    );
}
exports.isDoNotSetCallbackForSharePointMobileKSActive = isDoNotSetCallbackForSharePointMobileKSActive;
function isUseAADTenantUrlFromContextKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('db96e414-228c-4681-a9c9-f061e3fdc211'
    /* '04/18/2024', 'Use AAD tenant URL from context' */
    );
}
exports.isUseAADTenantUrlFromContextKSActivated = isUseAADTenantUrlFromContextKSActivated;
function isGuestUserForAnonymousAndEmailAuthenticatedKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('eae3e908-eb62-4f82-bdfb-88119e4e66dd'
    /* '04/29/2024', 'Treat both anonymous guest users as well as email authenticated guest users the same' */
    );
}
exports.isGuestUserForAnonymousAndEmailAuthenticatedKSActive = isGuestUserForAnonymousAndEmailAuthenticatedKSActive;
