// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/RecentFilesTab/RecentFilesTab.module.scss.js
require("./RecentFilesTab.module.css");
const styles = {
    filePicker: 'filePicker_7f8e8de5',
    tabsContainer: 'tabsContainer_7f8e8de5',
    tabContainer: 'tabContainer_7f8e8de5',
    nav: 'nav_7f8e8de5',
    tabHeaderContainer: 'tabHeaderContainer_7f8e8de5',
    breadcrumbNav: 'breadcrumbNav_7f8e8de5',
    breadcrumbNavItem: 'breadcrumbNavItem_7f8e8de5',
    tab: 'tab_7f8e8de5',
    tabOffset: 'tabOffset_7f8e8de5',
    tabFiles: 'tabFiles_7f8e8de5',
    scrollablePaneWrapper: 'scrollablePaneWrapper_7f8e8de5',
    actionButtonsContainer: 'actionButtonsContainer_7f8e8de5',
    actionButtons: 'actionButtons_7f8e8de5',
    actionButton: 'actionButton_7f8e8de5',
    tabHeader: 'tabHeader_7f8e8de5',
    selectedTab: 'selectedTab_7f8e8de5',
    header: 'header_7f8e8de5',
    xlg: 'xlg_7f8e8de5',
    xxlg: 'xxlg_7f8e8de5',
    xxxlg: 'xxxlg_7f8e8de5',
    xxxxlg: 'xxxxlg_7f8e8de5',
    itemPickerTopBar: 'itemPickerTopBar_7f8e8de5',
    commandBarNoChevron: 'commandBarNoChevron_7f8e8de5',
    fileItem: 'fileItem_7f8e8de5',
    folderItem: 'folderItem_7f8e8de5',
    fileTypeIcon: 'fileTypeIcon_7f8e8de5',
    fileTypeIconIcon: 'fileTypeIconIcon_7f8e8de5',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_7f8e8de5',
    iconColumnHeader: 'iconColumnHeader_7f8e8de5',
    fileRow: 'fileRow_7f8e8de5',
    folderRow: 'folderRow_7f8e8de5',
    emptyFolder: 'emptyFolder_7f8e8de5',
    emptyFolderImage: 'emptyFolderImage_7f8e8de5',
    emptyFolderImageTag: 'emptyFolderImageTag_7f8e8de5',
    emptyFolderTitle: 'emptyFolderTitle_7f8e8de5',
    emptyFolderSubText: 'emptyFolderSubText_7f8e8de5',
    emptyFolderMobile: 'emptyFolderMobile_7f8e8de5',
    emptyFolderPc: 'emptyFolderPc_7f8e8de5',
    emptyFolderIcon: 'emptyFolderIcon_7f8e8de5',
    gridListCell: 'gridListCell_7f8e8de5',
    cellRight: 'cellRight_7f8e8de5',
    cellLastRow: 'cellLastRow_7f8e8de5',
    itemTile: 'itemTile_7f8e8de5',
    isFile: 'isFile_7f8e8de5',
    isPhoto: 'isPhoto_7f8e8de5',
    isVideo: 'isVideo_7f8e8de5',
    hasThumbnail: 'hasThumbnail_7f8e8de5',
    itemTileContent: 'itemTileContent_7f8e8de5',
    itemTileDiscoverFileContainer: 'itemTileDiscoverFileContainer_7f8e8de5',
    itemTileFileContainer: 'itemTileFileContainer_7f8e8de5',
    itemTileFile: 'itemTileFile_7f8e8de5',
    itemTileThumbnail: 'itemTileThumbnail_7f8e8de5',
    image: 'image_7f8e8de5',
    itemTileCheckCircle: 'itemTileCheckCircle_7f8e8de5',
    isSelected: 'isSelected_7f8e8de5',
    itemTileNamePlate: 'itemTileNamePlate_7f8e8de5',
    itemTileName: 'itemTileName_7f8e8de5',
    itemTileChildCount: 'itemTileChildCount_7f8e8de5',
    itemTileSubText: 'itemTileSubText_7f8e8de5',
    recentGridList: 'recentGridList_7f8e8de5'
};
export default styles;
