// @microsoft/rush-lib/lib-commonjs/pluginFramework/PhasedCommandHooks.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.PhasedCommandHooks = void 0;
const tapable_1 = require("tapable");
/**
 * Hooks into the execution process for phased commands
 * @alpha
 */
class PhasedCommandHooks {
    constructor() {
        /**
         * Hook invoked to create operations for execution.
         * Use the context to distinguish between the initial run and phased runs.
         */
        this.createOperations = new tapable_1.AsyncSeriesWaterfallHook(['operations', 'context'], 'createOperations');
        /**
         * Hook invoked before operation start
         * Hook is series for stable output.
         */
        this.beforeExecuteOperations = new tapable_1.AsyncSeriesHook(['records', 'context']);
        /**
         * Hook invoked when operation status changed
         * Hook is series for stable output.
         */
        this.onOperationStatusChanged = new tapable_1.SyncHook(['record']);
        /**
         * Hook invoked after executing a set of operations.
         * Use the context to distinguish between the initial run and phased runs.
         * Hook is series for stable output.
         */
        this.afterExecuteOperations = new tapable_1.AsyncSeriesHook(['results', 'context']);
        /**
         * Hook invoked before executing a operation.
         */
        this.beforeExecuteOperation = new tapable_1.AsyncSeriesBailHook(['runnerContext'], 'beforeExecuteOperation');
        /**
         * Hook invoked to define environment variables for an operation.
         * May be invoked by the runner to get the environment for the operation.
         */
        this.createEnvironmentForOperation = new tapable_1.SyncWaterfallHook(['environment', 'runnerContext'], 'createEnvironmentForOperation');
        /**
         * Hook invoked after executing a operation.
         */
        this.afterExecuteOperation = new tapable_1.AsyncSeriesHook(['runnerContext'], 'afterExecuteOperation');
        /**
         * Hook invoked to shutdown long-lived work in plugins.
         */
        this.shutdownAsync = new tapable_1.AsyncParallelHook(undefined, 'shutdown');
        /**
         * Hook invoked after a run has finished and the command is watching for changes.
         * May be used to display additional relevant data to the user.
         * Only relevant when running in watch mode.
         */
        this.waitingForChanges = new tapable_1.SyncHook(undefined, 'waitingForChanges');
        /**
         * Hook invoked after executing operations and before waitingForChanges. Allows the caller
         * to augment or modify the log entry about to be written.
         */
        this.beforeLog = new tapable_1.SyncHook(['telemetryData'], 'beforeLog');
    }
}
exports.PhasedCommandHooks = PhasedCommandHooks;
