// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/DocumentTile/DocumentTile.module.scss.js
require("./DocumentTile.module.css");
const styles = {
    filePicker: 'filePicker_d0d16336',
    tabsContainer: 'tabsContainer_d0d16336',
    tabContainer: 'tabContainer_d0d16336',
    nav: 'nav_d0d16336',
    tabHeaderContainer: 'tabHeaderContainer_d0d16336',
    breadcrumbNav: 'breadcrumbNav_d0d16336',
    breadcrumbNavItem: 'breadcrumbNavItem_d0d16336',
    tab: 'tab_d0d16336',
    tabOffset: 'tabOffset_d0d16336',
    tabFiles: 'tabFiles_d0d16336',
    scrollablePaneWrapper: 'scrollablePaneWrapper_d0d16336',
    actionButtonsContainer: 'actionButtonsContainer_d0d16336',
    actionButtons: 'actionButtons_d0d16336',
    actionButton: 'actionButton_d0d16336',
    tabHeader: 'tabHeader_d0d16336',
    selectedTab: 'selectedTab_d0d16336',
    header: 'header_d0d16336',
    xlg: 'xlg_d0d16336',
    xxlg: 'xxlg_d0d16336',
    xxxlg: 'xxxlg_d0d16336',
    xxxxlg: 'xxxxlg_d0d16336',
    itemPickerTopBar: 'itemPickerTopBar_d0d16336',
    commandBarNoChevron: 'commandBarNoChevron_d0d16336',
    fileItem: 'fileItem_d0d16336',
    folderItem: 'folderItem_d0d16336',
    fileTypeIcon: 'fileTypeIcon_d0d16336',
    fileTypeIconIcon: 'fileTypeIconIcon_d0d16336',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_d0d16336',
    iconColumnHeader: 'iconColumnHeader_d0d16336',
    fileRow: 'fileRow_d0d16336',
    folderRow: 'folderRow_d0d16336',
    emptyFolder: 'emptyFolder_d0d16336',
    emptyFolderImage: 'emptyFolderImage_d0d16336',
    emptyFolderImageTag: 'emptyFolderImageTag_d0d16336',
    emptyFolderTitle: 'emptyFolderTitle_d0d16336',
    emptyFolderSubText: 'emptyFolderSubText_d0d16336',
    emptyFolderMobile: 'emptyFolderMobile_d0d16336',
    emptyFolderPc: 'emptyFolderPc_d0d16336',
    emptyFolderIcon: 'emptyFolderIcon_d0d16336',
    aboveNameplate: 'aboveNameplate_d0d16336',
    tileContent: 'tileContent_d0d16336',
    foreground: 'foreground_d0d16336',
    grid: 'grid_d0d16336',
    cellContent: 'cellContent_d0d16336',
    content: 'content_d0d16336',
    tile: 'tile_d0d16336',
    invokable: 'invokable_d0d16336',
    namePlate: 'namePlate_d0d16336',
    nameplate: 'nameplate_d0d16336',
    name: 'name_d0d16336',
    isLarge: 'isLarge_d0d16336',
    activity: 'activity_d0d16336',
    signalField: 'signalField_d0d16336',
    compact: 'compact_d0d16336',
    signalFieldValue: 'signalFieldValue_d0d16336',
    folderList: 'folderList_d0d16336',
    link: 'link_d0d16336',
    cell: 'cell_d0d16336',
    listCell: 'listCell_d0d16336',
    label: 'label_d0d16336',
    check: 'check_d0d16336',
    selectable: 'selectable_d0d16336',
    showCheck: 'showCheck_d0d16336',
    selected: 'selected_d0d16336',
    dialogMainOverride: 'dialogMainOverride_d0d16336',
    itemTileBlankCover: 'itemTileBlankCover_d0d16336',
    metadata: 'metadata_d0d16336',
    isSmall: 'isSmall_d0d16336',
    listPage: 'listPage_d0d16336',
    odItemTile2FolderCover: 'odItemTile2FolderCover_d0d16336',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_d0d16336',
    odItemTile2Image: 'odItemTile2Image_d0d16336',
    odImageFrame2: 'odImageFrame2_d0d16336',
    odImageFrame2Image: 'odImageFrame2Image_d0d16336',
    odImageFrame: 'odImageFrame_d0d16336',
    odImageStack: 'odImageStack_d0d16336',
    odImageStackTile: 'odImageStackTile_d0d16336',
    odImageTile: 'odImageTile_d0d16336',
    odImageTileBackground: 'odImageTileBackground_d0d16336',
    odImageTileImage: 'odImageTileImage_d0d16336',
    odItemTile2SmallIcon: 'odItemTile2SmallIcon_d0d16336'
};
export default styles;
