// @microsoft/spfx-heft-plugins/lib-commonjs/plugins/PluginNames.js
/**
 * @public
 */ "use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    CustomizeWebpackConfigPlugin: function() {
        return CustomizeWebpackConfigPlugin;
    },
    DeployAzureStoragePlugin: function() {
        return DeployAzureStoragePlugin;
    },
    DevDeployPlugin: function() {
        return DevDeployPlugin;
    },
    PackageSolutionPlugin: function() {
        return PackageSolutionPlugin;
    },
    SpfxHeftExtensionsPlugin: function() {
        return SpfxHeftExtensionsPlugin;
    },
    SpfxHeftPlugin: function() {
        return SpfxHeftPlugin;
    },
    SpfxWebpackConfigurationPlugin: function() {
        return SpfxWebpackConfigurationPlugin;
    },
    SpfxWebpackServePlugin: function() {
        return SpfxWebpackServePlugin;
    },
    ThirdPartyExternalsPlugin: function() {
        return ThirdPartyExternalsPlugin;
    }
});
const DevDeployPlugin = 'spfx-dev-deploy-plugin';
const SpfxHeftPlugin = 'spfx-heft-plugin';
const SpfxHeftExtensionsPlugin = 'spfx-heft-extensions-plugin';
const SpfxWebpackConfigurationPlugin = 'spfx-webpack-configuration-plugin';
const SpfxWebpackServePlugin = 'spfx-webpack-serve-plugin';
const CustomizeWebpackConfigPlugin = 'customize-spfx-webpack-configuration-plugin';
const PackageSolutionPlugin = 'package-solution-plugin';
const ThirdPartyExternalsPlugin = 'third-party-externals-plugin';
const DeployAzureStoragePlugin = 'deploy-azure-storage-plugin';

//#sourceMappingUrl=./PluginNames.js.map
