// @microsoft/sp-webpart-base/lib-commonjs/common/KillSwitches.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.isClickToAddFromContentPaneKSActivated = exports.isTranspileAlreadyLoadedWPKSActivated = exports.isFixUnstableThumbnailKSActivated = exports.isHandleDisposeRaceConditionKSActive = exports.isRefactorLoadWebPartKSActivated = exports.isFixUnwantedAudienceTargetingForExpandedPropertyPaneKSActivated = exports.updateMissingWebPartLifeCycleVetoesKSActivated = exports.isAudienceTargetingForMultiColumnPropertyPaneKSActive = exports.isShowErrorAceForBlockedAceKSActivated = exports.isOverrideDataPersistedFlagKSActive = exports.isUsePromiseAllForFetchWebPartsKSActivated = exports.isImageRefactorKSActivated = exports.isAsyncCompleteKSActivated = exports.isWebPartManifestListCacheRefreshKSActivated = exports.isAccessibleRootKSActivated = exports.isDashboardWebPartMediumCardBreakPointKSActivated = exports.isAddMoreRenderErrorInfoKSActivated = exports.isDashboardWebPartVerticalLayoutCardWidthFix = exports.isExemptWebPartWithDynamicDataFromLazyLoadKSActivated = exports.isAlwaysShowIsolatedACEOnTeamsKSActive = exports.isVivaHomeIsolatedAceCheckKSActive = exports.isPropertyPaneUpdateASHAVetoKSActivated = exports.isIsolatedACEWPRenderAsyncKSActivated = exports.isPreventLateIframeRedirectKSActivated = exports.isUberIsolatedACEFlightKSActivated = exports.isSafariIsolationKSActivated = exports.isLogWebPartLoadErrorKSActivated = exports.isFixMissingContextKSActivated = exports.isDynamicIframedPropertyPaneHandlingKSActivated = exports.isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated = exports.isChainDisposeCallsAfterOninitKSActivated = exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = exports.isSpartanListWebPartExpKSActivated = exports.isConfigurationToolsKSActivated = exports.isIsolatedWebPartTabStopKSActivated = void 0;
var sp_core_library_1 = require("@microsoft/sp-core-library");
var KillSwitches = /** @class */ (function () {
    function KillSwitches() {
    }
    return KillSwitches;
}());
exports.default = KillSwitches;
function isIsolatedWebPartTabStopKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d8618918-ac55-4979-b9d4-be944a4feb95'
    /* '10/25/2021', 'adds tabindex=0 to the isolated webpart iframe' */
    );
}
exports.isIsolatedWebPartTabStopKSActivated = isIsolatedWebPartTabStopKSActivated;
// cross-project kill switch
function isConfigurationToolsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('a3f45048-0db5-4b7c-9a9a-e3183d216c4d'
    /** '03/21/2024', 'Enables configuration tools pivot in content panel. */
    );
}
exports.isConfigurationToolsKSActivated = isConfigurationToolsKSActivated;
function isSpartanListWebPartExpKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('d0f08b8b-edb0-4056-81ca-75732f5cdcbd'
    /** '09/18/2024', 'erabelle: Spartan list webpart experiment' */
    );
}
exports.isSpartanListWebPartExpKSActivated = isSpartanListWebPartExpKSActivated;
function isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b8c50ded-d048-4eb9-bc41-21190d9d3bbb'
    /* '01/20/2022', 'Work around to enable property pane for isolated ACE web' */
    );
}
exports.isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated = isIsolatedAdaptiveCardExtensionWebPropertyPaneWorkAroundKSActivated;
function isChainDisposeCallsAfterOninitKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ca984138-b518-4cf1-b201-2efe9322bb5e'
    /* '02/01/2022', 'Ensures that calls to dispose always wait for any activing onInit calls' */
    );
}
exports.isChainDisposeCallsAfterOninitKSActivated = isChainDisposeCallsAfterOninitKSActivated;
function isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('ab6a720f-9915-4635-9b30-71b5d83e0647'
    /* '01/21/2022', 'Creates a functional iframed property pane for isolated ACEs' */
    );
}
exports.isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated = isIsolatedAdaptiveCardExtensionIframedPropertyPaneKSActivated;
function isDynamicIframedPropertyPaneHandlingKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('1a2da4d0-9678-4a79-95a4-f0b9a7732dc2'
    /* '02/02/2022', 'Dynamically creates and assigns an iframed property pane to selected IBaseFramedWebParts' */
    );
}
exports.isDynamicIframedPropertyPaneHandlingKSActivated = isDynamicIframedPropertyPaneHandlingKSActivated;
function isFixMissingContextKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e1e44790-a263-4f43-8b40-5b5e9b170300'
    /* '02/22/2022', 'Fix missing context when passing down to propertypane' */
    );
}
exports.isFixMissingContextKSActivated = isFixMissingContextKSActivated;
function isLogWebPartLoadErrorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('c041bc0e-fcb5-4a3e-a4bc-d65a00a39a86'
    /* '02/24/2022', 'Check if exception thrown during loadWebPart is an error and ' */
    );
}
exports.isLogWebPartLoadErrorKSActivated = isLogWebPartLoadErrorKSActivated;
function isSafariIsolationKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('61a9578e-c6d2-4605-970f-222a8b29196c' /* '03/13/22', 'Skip rendering isolated ACEs in Safari' */);
}
exports.isSafariIsolationKSActivated = isSafariIsolationKSActivated;
function isUberIsolatedACEFlightKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('07684c68-2e8e-4a25-b7e0-330781a92661'
    /* '03/16/22', 'If the KS is not pulled check the uber IsolatedAdaptiveCardExtensionWeb flight otherwise check APEXIsolatedACEWEB flights' */
    );
}
exports.isUberIsolatedACEFlightKSActivated = isUberIsolatedACEFlightKSActivated;
function isPreventLateIframeRedirectKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('56151a3e-a100-403c-9e91-cb5d1677ed66'
    /* '06/13/2022', 'instantiate IframeLoadRedirector earlier and support pre-emptively preventing redirects' */
    );
}
exports.isPreventLateIframeRedirectKSActivated = isPreventLateIframeRedirectKSActivated;
function isIsolatedACEWPRenderAsyncKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e8edc935-7cd4-463b-81f5-aa1cce74d7d0'
    /* '06/27/2022', 'Captures if async web part has finished rendering using LKG or updated content.' */
    );
}
exports.isIsolatedACEWPRenderAsyncKSActivated = isIsolatedACEWPRenderAsyncKSActivated;
function isPropertyPaneUpdateASHAVetoKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('21f00b7b-3ed6-4c3a-bd49-8639460de07e'
    /* '08/02/2022', 'Add monitor to capture property pane update scenario for ASHA' */
    );
}
exports.isPropertyPaneUpdateASHAVetoKSActivated = isPropertyPaneUpdateASHAVetoKSActivated;
function isVivaHomeIsolatedAceCheckKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('f993c5d6-5637-47bc-802d-6da6dda11a27' /* '08/24/22', 'Fix the isolated ACE flight check for Viva Home' */);
}
exports.isVivaHomeIsolatedAceCheckKSActive = isVivaHomeIsolatedAceCheckKSActive;
function isAlwaysShowIsolatedACEOnTeamsKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('6c457592-8c32-472e-b340-349c014d6bc2' /* '10/20/2022', 'Always show isolated cards on Teams' */);
}
exports.isAlwaysShowIsolatedACEOnTeamsKSActive = isAlwaysShowIsolatedACEOnTeamsKSActive;
function isExemptWebPartWithDynamicDataFromLazyLoadKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('bbd78e03-ff18-40ed-9a94-973fc659366c' /* '12/01/2022', 'Update getWebPartsExemptedFromViewPortLoading to check for containsDynamicDataSource' */);
}
exports.isExemptWebPartWithDynamicDataFromLazyLoadKSActivated = isExemptWebPartWithDynamicDataFromLazyLoadKSActivated;
function isDashboardWebPartVerticalLayoutCardWidthFix() {
    return sp_core_library_1._SPKillSwitch.isActivated('43ce53d4-53da-48c4-acbd-20ff31eff5ee' /* '1/09/2023', 'Fix medium card width for dashboard web part in vertical layout.' */);
}
exports.isDashboardWebPartVerticalLayoutCardWidthFix = isDashboardWebPartVerticalLayoutCardWidthFix;
function isAddMoreRenderErrorInfoKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('290c0d46-6ee0-4d0b-aee5-ef3f01e4020e' /* '01/11/2023', 'Add more RenderError Information' */);
}
exports.isAddMoreRenderErrorInfoKSActivated = isAddMoreRenderErrorInfoKSActivated;
function isDashboardWebPartMediumCardBreakPointKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('8b372850-6897-4ef2-aecd-cd615019e544' /* '03/06/2023', 'Reduce size of medium cards if we cannot fit at least two min width medium cards in a row for dashboard web part.' */);
}
exports.isDashboardWebPartMediumCardBreakPointKSActivated = isDashboardWebPartMediumCardBreakPointKSActivated;
function isAccessibleRootKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('33558902-7587-4229-a570-633976fa332b' /* '03/28/2023', 'Change the DOM hierarchy of where the accessible web part label is rendered' */);
}
exports.isAccessibleRootKSActivated = isAccessibleRootKSActivated;
function isWebPartManifestListCacheRefreshKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('98bb76c9-49c4-4736-80b9-88ff41d65cac' /* '04/24/2023', 'Refreshes the webpart cache list after loading the new manifest list' */);
}
exports.isWebPartManifestListCacheRefreshKSActivated = isWebPartManifestListCacheRefreshKSActivated;
function isAsyncCompleteKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6e0a6d0e-ec0c-4667-85ce-198b4bbe78af' /* '05/17/2023', 'Allow render complete async callback to work if called quickly' */);
}
exports.isAsyncCompleteKSActivated = isAsyncCompleteKSActivated;
function isImageRefactorKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('138ce21a-eef4-4e30-b350-4225400c72ac'
    /* 05/30/2023, 'Add KS to disable new image webpart experience' */
    );
}
exports.isImageRefactorKSActivated = isImageRefactorKSActivated;
function isUsePromiseAllForFetchWebPartsKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('69933c68-aef0-4f8f-8016-a5974d22b2c6'
    /* '12/08/2023', 'Use Promise.all for fetchWebParts' */
    );
}
exports.isUsePromiseAllForFetchWebPartsKSActivated = isUsePromiseAllForFetchWebPartsKSActivated;
function isOverrideDataPersistedFlagKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('bf6ec07d-4440-4ee6-83c6-17b560599c30'
    /* '2/21/2024', 'Allow consumers of MinimalWebPartContainer to override flag for whether data is from persisted store' */
    );
}
exports.isOverrideDataPersistedFlagKSActive = isOverrideDataPersistedFlagKSActive;
function isShowErrorAceForBlockedAceKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('2087be6f-d433-41f5-b91a-5416f455d74d'
    /* '01/19/2024', 'Show Error ACE for blocked ACE' */
    );
}
exports.isShowErrorAceForBlockedAceKSActivated = isShowErrorAceForBlockedAceKSActivated;
function isAudienceTargetingForMultiColumnPropertyPaneKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('4f5fdd22-e54b-443e-b79a-6134a5e59e8d'
    /* '03/19/2024', 'Add configuration option for audience targeting in multi column property pane' */
    );
}
exports.isAudienceTargetingForMultiColumnPropertyPaneKSActive = isAudienceTargetingForMultiColumnPropertyPaneKSActive;
function updateMissingWebPartLifeCycleVetoesKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('4a64c8cf-892b-4f83-8fe4-3a2245297f76'
    /* '03/21/2024', 'atagle - update missing web part lifecycle vetoes' */
    );
}
exports.updateMissingWebPartLifeCycleVetoesKSActivated = updateMissingWebPartLifeCycleVetoesKSActivated;
function isFixUnwantedAudienceTargetingForExpandedPropertyPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('6d1353bf-844b-4683-afdd-07d4dae51ea2'
    /* '05/17/2024', 'Ensure only add audience targeting field if explicitly configured for expanded property pane' */
    );
}
exports.isFixUnwantedAudienceTargetingForExpandedPropertyPaneKSActivated = isFixUnwantedAudienceTargetingForExpandedPropertyPaneKSActivated;
function isRefactorLoadWebPartKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e11d11a3-f2a0-4cc3-8c36-60f157abdb47'
    /* '08/01/2024', 'Refactor load webpart to separate render' */
    );
}
exports.isRefactorLoadWebPartKSActivated = isRefactorLoadWebPartKSActivated;
function isHandleDisposeRaceConditionKSActive() {
    return sp_core_library_1._SPKillSwitch.isActivated('a1b3f56e-4ebf-4ae2-bb75-146286e336e4'
    /* '10/21/2024', 'liaye - Handle race condition that caused by disposed web parts' */
    );
}
exports.isHandleDisposeRaceConditionKSActive = isHandleDisposeRaceConditionKSActive;
function isFixUnstableThumbnailKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('e34f2e15-eaed-404b-93ee-bed24f7e5eec' /* '10/17/2024', 'tialu: fix unstable thumbnail' */);
}
exports.isFixUnstableThumbnailKSActivated = isFixUnstableThumbnailKSActivated;
function isTranspileAlreadyLoadedWPKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('b56be5bc-975c-4c10-b387-808c5404f3b3'
    /* '11/04/2024', 'poorvitusam - Use loaded webpart for transpile' */
    );
}
exports.isTranspileAlreadyLoadedWPKSActivated = isTranspileAlreadyLoadedWPKSActivated;
function isClickToAddFromContentPaneKSActivated() {
    return sp_core_library_1._SPKillSwitch.isActivated('757e6220-8354-49d2-979a-cd70080da6d9' /* '10/29/2024', 'jayharrs click to add from content panel to existing web parts' */);
}
exports.isClickToAddFromContentPaneKSActivated = isClickToAddFromContentPaneKSActivated;
