// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/FileBrowser/FileBrowser.module.scss.js
require("./FileBrowser.module.css");
const styles = {
    filePicker: 'filePicker_9826e513',
    tabsContainer: 'tabsContainer_9826e513',
    tabContainer: 'tabContainer_9826e513',
    nav: 'nav_9826e513',
    tabHeaderContainer: 'tabHeaderContainer_9826e513',
    breadcrumbNav: 'breadcrumbNav_9826e513',
    breadcrumbNavItem: 'breadcrumbNavItem_9826e513',
    tab: 'tab_9826e513',
    tabOffset: 'tabOffset_9826e513',
    tabFiles: 'tabFiles_9826e513',
    scrollablePaneWrapper: 'scrollablePaneWrapper_9826e513',
    actionButtonsContainer: 'actionButtonsContainer_9826e513',
    actionButtons: 'actionButtons_9826e513',
    actionButton: 'actionButton_9826e513',
    tabHeader: 'tabHeader_9826e513',
    selectedTab: 'selectedTab_9826e513',
    header: 'header_9826e513',
    xlg: 'xlg_9826e513',
    xxlg: 'xxlg_9826e513',
    xxxlg: 'xxxlg_9826e513',
    xxxxlg: 'xxxxlg_9826e513',
    itemPickerTopBar: 'itemPickerTopBar_9826e513',
    commandBarNoChevron: 'commandBarNoChevron_9826e513',
    fileItem: 'fileItem_9826e513',
    folderItem: 'folderItem_9826e513',
    fileTypeIcon: 'fileTypeIcon_9826e513',
    fileTypeIconIcon: 'fileTypeIconIcon_9826e513',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_9826e513',
    iconColumnHeader: 'iconColumnHeader_9826e513',
    fileRow: 'fileRow_9826e513',
    folderRow: 'folderRow_9826e513',
    emptyFolder: 'emptyFolder_9826e513',
    emptyFolderImage: 'emptyFolderImage_9826e513',
    emptyFolderImageTag: 'emptyFolderImageTag_9826e513',
    emptyFolderTitle: 'emptyFolderTitle_9826e513',
    emptyFolderSubText: 'emptyFolderSubText_9826e513',
    emptyFolderMobile: 'emptyFolderMobile_9826e513',
    emptyFolderPc: 'emptyFolderPc_9826e513',
    emptyFolderIcon: 'emptyFolderIcon_9826e513'
};
export default styles;
