// @microsoft/eslint-plugin-spfx/lib/no-require-ensure.js
"use strict";
// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.
Object.defineProperty(exports, "__esModule", { value: true });
exports.noRequireEnsureRule = exports.MESSAGE_ID = void 0;
const utils_1 = require("@typescript-eslint/utils");
exports.MESSAGE_ID = 'error-require-ensure';
const noRequireEnsureRule = {
    defaultOptions: [],
    meta: {
        type: 'problem',
        messages: {
            [exports.MESSAGE_ID]: 'Usage of "require.ensure(...)" is outdated. Use "import(...)" instead.'
        },
        schema: [],
        docs: {
            description: 'Prevent usage of Webpack\'s "require.ensure" method in favor of "import"',
            url: 'https://www.npmjs.com/package/@rushstack/eslint-plugin'
        }
    },
    create: (context) => {
        return {
            CallExpression: (node) => {
                if (node.callee.type === utils_1.TSESTree.AST_NODE_TYPES.MemberExpression) {
                    const { object, property } = node.callee;
                    if (object.type === utils_1.TSESTree.AST_NODE_TYPES.Identifier && object.name === 'require') {
                        if ((property.type === utils_1.TSESTree.AST_NODE_TYPES.Identifier && property.name === 'ensure') ||
                            (property.type === utils_1.TSESTree.AST_NODE_TYPES.Literal && property.value === 'ensure')) {
                            context.report({ node, messageId: exports.MESSAGE_ID });
                        }
                    }
                }
            }
        };
    }
};
exports.noRequireEnsureRule = noRequireEnsureRule;
