// @pnp/spfx-property-controls/lib/propertyFields/filePicker/filePickerControls/controls/TilesList/TilesList.module.scss.js
require("./TilesList.module.css");
const styles = {
    filePicker: 'filePicker_1159f4e1',
    tabsContainer: 'tabsContainer_1159f4e1',
    tabContainer: 'tabContainer_1159f4e1',
    nav: 'nav_1159f4e1',
    tabHeaderContainer: 'tabHeaderContainer_1159f4e1',
    breadcrumbNav: 'breadcrumbNav_1159f4e1',
    breadcrumbNavItem: 'breadcrumbNavItem_1159f4e1',
    tab: 'tab_1159f4e1',
    tabOffset: 'tabOffset_1159f4e1',
    tabFiles: 'tabFiles_1159f4e1',
    scrollablePaneWrapper: 'scrollablePaneWrapper_1159f4e1',
    actionButtonsContainer: 'actionButtonsContainer_1159f4e1',
    actionButtons: 'actionButtons_1159f4e1',
    actionButton: 'actionButton_1159f4e1',
    tabHeader: 'tabHeader_1159f4e1',
    selectedTab: 'selectedTab_1159f4e1',
    header: 'header_1159f4e1',
    xlg: 'xlg_1159f4e1',
    xxlg: 'xxlg_1159f4e1',
    xxxlg: 'xxxlg_1159f4e1',
    xxxxlg: 'xxxxlg_1159f4e1',
    itemPickerTopBar: 'itemPickerTopBar_1159f4e1',
    commandBarNoChevron: 'commandBarNoChevron_1159f4e1',
    fileItem: 'fileItem_1159f4e1',
    folderItem: 'folderItem_1159f4e1',
    fileTypeIcon: 'fileTypeIcon_1159f4e1',
    fileTypeIconIcon: 'fileTypeIconIcon_1159f4e1',
    detailsListHeaderFileTypeIcon: 'detailsListHeaderFileTypeIcon_1159f4e1',
    iconColumnHeader: 'iconColumnHeader_1159f4e1',
    fileRow: 'fileRow_1159f4e1',
    folderRow: 'folderRow_1159f4e1',
    emptyFolder: 'emptyFolder_1159f4e1',
    emptyFolderImage: 'emptyFolderImage_1159f4e1',
    emptyFolderImageTag: 'emptyFolderImageTag_1159f4e1',
    emptyFolderTitle: 'emptyFolderTitle_1159f4e1',
    emptyFolderSubText: 'emptyFolderSubText_1159f4e1',
    emptyFolderMobile: 'emptyFolderMobile_1159f4e1',
    emptyFolderPc: 'emptyFolderPc_1159f4e1',
    emptyFolderIcon: 'emptyFolderIcon_1159f4e1',
    aboveNameplate: 'aboveNameplate_1159f4e1',
    tileContent: 'tileContent_1159f4e1',
    foreground: 'foreground_1159f4e1',
    grid: 'grid_1159f4e1',
    cellContent: 'cellContent_1159f4e1',
    content: 'content_1159f4e1',
    tile: 'tile_1159f4e1',
    invokable: 'invokable_1159f4e1',
    namePlate: 'namePlate_1159f4e1',
    nameplate: 'nameplate_1159f4e1',
    name: 'name_1159f4e1',
    isLarge: 'isLarge_1159f4e1',
    activity: 'activity_1159f4e1',
    signalField: 'signalField_1159f4e1',
    compact: 'compact_1159f4e1',
    signalFieldValue: 'signalFieldValue_1159f4e1',
    folderList: 'folderList_1159f4e1',
    link: 'link_1159f4e1',
    cell: 'cell_1159f4e1',
    listCell: 'listCell_1159f4e1',
    label: 'label_1159f4e1',
    check: 'check_1159f4e1',
    selectable: 'selectable_1159f4e1',
    showCheck: 'showCheck_1159f4e1',
    selected: 'selected_1159f4e1',
    dialogMainOverride: 'dialogMainOverride_1159f4e1',
    itemTileBlankCover: 'itemTileBlankCover_1159f4e1',
    metadata: 'metadata_1159f4e1',
    isSmall: 'isSmall_1159f4e1',
    listPage: 'listPage_1159f4e1',
    odItemTile2FolderCover: 'odItemTile2FolderCover_1159f4e1',
    standaloneListBreadcrumb: 'standaloneListBreadcrumb_1159f4e1'
};
export default styles;
