// @microsoft/sp-property-pane/lib-commonjs/propertyPaneFields/propertyPaneWebpartTitle/getPropertyPaneHeadingLevelDropdownProperties.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.getPropertyPaneHeadingLevelDropdownProperties = void 0;
var tslib_1 = require("tslib");
var react_1 = tslib_1.__importDefault(require("react"));
var sp_core_library_1 = require("@microsoft/sp-core-library");
var office_ui_fabric_react_bundle_1 = require("@ms/office-ui-fabric-react-bundle");
var sp_heading_level_1 = require("@ms/sp-heading-level");
var Flights_1 = require("../../common/Flights");
var KillSwitches_1 = require("../../common/KillSwitches");
var Strings_resx_1 = tslib_1.__importDefault(require("../../loc/Strings.resx"));
/**
 * Helper method to get an Heading level dropdown properties on the PropertyPane.
 *
 * @internal
 */
function getPropertyPaneHeadingLevelDropdownProperties(headingLevel, moduleName, dataAutomationId, controlName) {
    return {
        label: Strings_resx_1.default.headingLevelLabel,
        ariaLabel: Strings_resx_1.default.headingLevelLabel,
        dataAutomationId: dataAutomationId,
        tooltipProps: {
            tooltipText: Strings_resx_1.default.headingLevelTooltipDesc
        },
        options: [
            {
                key: 2,
                text: !(0, KillSwitches_1.isLocalizeWebPartTitleHeadingLevelsKSActivated)()
                    ? sp_core_library_1.Text.format(Strings_resx_1.default.headingLevelOptionLabel, '2')
                    : 'Heading 2'
            },
            {
                key: 3,
                text: !(0, KillSwitches_1.isLocalizeWebPartTitleHeadingLevelsKSActivated)()
                    ? sp_core_library_1.Text.format(Strings_resx_1.default.headingLevelOptionLabel, '3')
                    : 'Heading 3'
            },
            {
                key: 4,
                text: !(0, KillSwitches_1.isLocalizeWebPartTitleHeadingLevelsKSActivated)()
                    ? sp_core_library_1.Text.format(Strings_resx_1.default.headingLevelOptionLabel, '4')
                    : 'Heading 4'
            }
        ],
        selectedKey: headingLevel ? headingLevel : sp_heading_level_1._DEFAULT_HEADING_LEVEL,
        onRenderOption: (0, Flights_1.isModifyFontSizeTitleHeadingsEnabled)() ? getHeadingLevelOption : undefined
    };
}
exports.getPropertyPaneHeadingLevelDropdownProperties = getPropertyPaneHeadingLevelDropdownProperties;
function getHeadingLevelOption(option) {
    if (option) {
        return (react_1.default.createElement("div", { style: { display: 'inline-flex' } },
            react_1.default.createElement(office_ui_fabric_react_bundle_1.Label, { style: {
                    fontSize: (0, sp_heading_level_1._HeadingLevel)(option.key).fontSize,
                    fontWeight: '$ms-font-weight-semibold'
                } }, option.text)));
    }
    return react_1.default.createElement(react_1.default.Fragment, null);
}
