// @microsoft/sp-module-interfaces/lib-commonjs/manifestSchemas/examples/application_2.manifest.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const appManifest = {
    manifestVersion: 2,
    componentType: 'Application',
    id: '00000000-0000-0000-0000-000000000000',
    alias: 'myApplication',
    version: '0.0.1',
    requiresCustomScript: false,
    loaderConfig: {
        internalModuleBaseUrls: [
            'https://cdn.net/'
        ],
        entryModuleId: 'main.bundle',
        exportName: 'foobar',
        scriptResources: {
            'main.bundle': {
                type: 'path',
                path: 'bundle.script.js'
            }
        }
    },
    preloadOptions: {
        shouldPreloadWeb: true,
        shouldPreloadUser: true,
        shouldPreloadList: true,
        shouldPreloadItem: true,
        shouldPreloadQuickLaunch: true,
        preloadListItemQueryOption: 'query',
        preloadListItemQueryOptionOverride: 'query',
        preloadListItemProperties: [
            'DefaultCanvasContentVersion'
        ]
    },
    preloadComponents: [
        'framework1',
        'framework2'
    ],
    title: {
        default: 'foobar'
    },
    description: {
        default: 'foobar baz'
    },
    hasSuiteNav: true
};
const _default = appManifest;

//#sourceMappingUrl=./application_2.manifest.js.map
